---
title: "G by lifestage interaction"
author: "Meng"
date: "2024-07-16"
output: html_document
---

```{r setup, include=FALSE}
library(dplyr)
library(ggplot2)
library(patchwork)
data <- "/Users/yuanmeng/Library/CloudStorage/OneDrive-UniversityofToronto/Manuscripts/rumex_eqtl/eqtl/old/"
```

# MAF plots
```{r}
FL_eqtl_full <- FL_eqtl_full %>% mutate(category = ifelse(pval_nominal < pval_nominal_threshold, "Significant", "Not_Significant"))
ML_eqtl_full <- ML_eqtl_full %>% mutate(category = ifelse(pval_nominal < pval_nominal_threshold, "Significant", "Not_Significant"))
MP_eqtl_full <- MP_eqtl_full %>% mutate(category = ifelse(pval_nominal < pval_nominal_threshold, "Significant", "Not_Significant"))
```

separate left joins
```{r}
ml <- ML_eqtl_sig_min %>% select(phenotype_id, variant_id, slope)
fl <- FL_eqtl_full %>% select(phenotype_id, variant_id, slope, category)

pair1 <- inner_join(ml, fl, by = c("phenotype_id","variant_id"))
colnames(pair1)[3:5] <- c("male_leaf", "female_leaf", "sig_fl") # 
pair1$sig_ml <- "Significant"

ml <- ML_eqtl_full %>% select(phenotype_id, variant_id, slope, category)
fl <- FL_eqtl_sig_min %>% select(phenotype_id, variant_id, slope)
pair2 <- inner_join(fl, ml, by = c("phenotype_id","variant_id"))
colnames(pair2)[3:5] <- c("female_leaf", "male_leaf", "sig_ml") # 
pair2$sig_fl <- "Significant"

ml_fl <- rbind(pair1, pair2)
ml_fl <- ml_fl %>%  mutate(category = case_when(
    sig_ml == "Significant" & sig_fl != "Significant" ~ "male",
    sig_fl == "Significant" & sig_ml != "Significant" ~ "female",
    sig_ml == "Significant" & sig_fl == "Significant" ~ "both"
    ))
```
 
```{r}
cor.test(ml_fl$male_leaf, ml_fl$female_leaf, method=c("spearman")) 
cor.test(ml_mp$male_leaf, ml_mp$pollen, method=c("spearman")) 
```
```{r}
rsq <- function(x, y) summary(lm(y~x))$r.squared
rsq(ml_mp$male_leaf, ml_mp$pollen)
rsq(ml_fl$male_leaf, ml_fl$female_leaf)
```
 
 
```{r}
ml <- ML_eqtl_sig_min %>% select(phenotype_id, variant_id, slope)
mp <- MP_eqtl_full %>% select(phenotype_id, variant_id, slope, category)
pair1 <- inner_join(ml, mp, by = c("phenotype_id","variant_id"))
colnames(pair1)[3:5] <- c("male_leaf", "pollen", "sig_mp") # 
pair1$sig_ml <- "Significant"

ml <- ML_eqtl_full %>% select(phenotype_id, variant_id, slope, category)
mp <- MP_eqtl_sig_min %>% select(phenotype_id, variant_id, slope)
pair2 <- inner_join(mp, ml, by = c("phenotype_id","variant_id"))
colnames(pair2)[3:5] <- c("pollen", "male_leaf", "sig_ml") # 
pair2$sig_mp <- "Significant"

ml_mp <- rbind(pair1, pair2)
ml_mp <- ml_mp %>%  mutate(category = case_when(
    sig_ml == "Significant" & sig_mp != "Significant" ~ "leaf",
    sig_mp == "Significant" & sig_ml != "Significant" ~ "pollen",
    sig_ml == "Significant" & sig_mp == "Significant" ~ "both"
    ))
```

```{r}
p1 <- ggplot(ml_fl, aes(x = male_leaf, y = female_leaf, color = category)) + geom_point(size=0.8)+ theme_classic() + labs(x="Male leaf", y="Female leaf") + theme(text = element_text(size = 11))+ scale_colour_manual(values = c("black", "#FFCCCB", "#ADD8E6"))+ geom_vline(xintercept=0, color = "black",linetype="dashed") + geom_hline(yintercept=0, color = "black",linetype="dashed")

p2 <- ggplot(ml_mp, aes(x = male_leaf, y = pollen, color = category)) + geom_point(size=0.8)+ theme_classic() + labs(x="Male leaf", y="Pollen") + theme(text = element_text(size = 11))+ scale_colour_manual(values = c("black", "#CBC3E3", "#FFDBBB")) + geom_vline(xintercept=0, color = "black",linetype="dashed") + geom_hline(yintercept=0, color = "black",linetype="dashed")

p1 + p2
```


# extract the off diagonal eqtls 
## ml
```{r}
ml <- ML_eqtl_sig_min %>% select(phenotype_id, variant_id, slope)
mp <- MP_eqtl_sig %>% select(phenotype_id, variant_id, slope)
ml_mp <- inner_join(ml, mp, by = c("phenotype_id","variant_id"))
colnames(ml_mp)[3:4] <- c("male_leaf", "pollen") # 234

ml_mp_inter1 <- ml_mp %>% filter(male_leaf > 0 & pollen < 0) # 4
ml_mp_inter2 <- ml_mp %>% filter(male_leaf < 0 & pollen > 0) # 8
ml_mp_inter <- rbind(ml_mp_inter1, ml_mp_inter2)
```

## fl
```{r}
fl <- FL_eqtl_sig_min %>% select(phenotype_id, variant_id, slope)
mp <- MP_eqtl_sig %>% select(phenotype_id, variant_id, slope)

fl_mp <- inner_join(fl, mp, by = c("phenotype_id","variant_id"))
colnames(fl_mp)[3:4] <- c("female_leaf", "pollen") # 89

fl_mp_inter1 <- fl_mp %>% filter(female_leaf > 0 & pollen < 0) # 2
fl_mp_inter2 <- fl_mp %>% filter(female_leaf < 0 & pollen > 0) # 8
fl_mp_inter <- rbind(fl_mp_inter1, fl_mp_inter2)
```

## mp
```{r}
ml <- ML_eqtl_sig %>% select(phenotype_id, variant_id, slope)
fl <- FL_eqtl_sig %>% select(phenotype_id, variant_id, slope)
mp <- MP_eqtl_sig_min %>% select(phenotype_id, variant_id, slope)

mp_ml <- inner_join(mp, ml, by = c("phenotype_id","variant_id"))
colnames(mp_ml)[3:4] <- c("pollen", "male_leaf") # 249
mp_fl <- inner_join(mp, fl, by = c("phenotype_id","variant_id"))
colnames(mp_fl)[3:4] <- c("pollen", "female_leaf") # 144

mp_ml_inter1 <- mp_ml %>% filter(female_leaf < 0 & pollen > 0) # 8
mp_ml_inter2 <- mp_ml %>% filter(female_leaf > 0 & pollen < 0) # 6
mp_ml_inter <- rbind(mp_ml_inter1, mp_ml_inter2) 

mp_fl_inter1 <- mp_fl %>% filter(female_leaf < 0 & pollen > 0) # 6
mp_fl_inter2 <- mp_fl %>% filter(female_leaf > 0 & pollen < 0) # 5
mp_fl_inter <- rbind(mp_fl_inter1, mp_fl_inter2)

write.table(mp_ml_inter$variant_id, file = paste0(data, "mp_ml_inter.txt"), row.names = F, quote = F, col.names = F)
write.table(mp_fl_inter$variant_id, file = paste0(data, "mp_fl_inter.txt"), row.names = F, quote = F, col.names = F)

write.table(ml_mp_inter$variant_id, file = paste0(data, "ml_mp_inter.txt"), row.names = F, quote = F, col.names = F)
write.table(fl_mp_inter$variant_id, file = paste0(data, "fl_mp_inter.txt"), row.names = F, quote = F, col.names = F)
```

, "maf","ma_samples", "ma_count"
```{r}
ml <- ML_eqtl_sig 
mp <- MP_eqtl_sig_min 
mp_ml <- inner_join(mp, ml, by = c("phenotype_id","variant_id", "start_distance")) # 249

#colnames(mp_ml)[3:4] <- c("pollen", "male_leaf") # 249

mp_ml_inter1 <- mp_ml %>% filter(slope.x < 0 & slope.y > 0) # 8
mp_ml_inter2 <- mp_ml %>% filter(slope.x > 0 & slope.y < 0) # 6
mp_ml_inter <- rbind(mp_ml_inter1, mp_ml_inter2) 
mp_ml_inter$top_in_pollen <- "yes"
mp_ml_inter$top_in_leaf <- "no"

ml <- ML_eqtl_sig_min 
mp <- MP_eqtl_sig 
ml_mp <- inner_join(mp, ml, by = c("phenotype_id","variant_id", "start_distance")) # 234

#colnames(ml_mp)[3:4] <- c("pollen", "male_leaf") # 249

ml_mp_inter1 <- ml_mp %>% filter(slope.x < 0 & slope.y > 0) # 8
ml_mp_inter2 <- ml_mp %>% filter(slope.x > 0 & slope.y < 0) # 4
ml_mp_inter <- rbind(ml_mp_inter1, ml_mp_inter2) # 12
ml_mp_inter$top_in_leaf <- "yes"
ml_mp_inter$top_in_pollen <- "no"

lifestage_inter <- rbind(ml_mp_inter, mp_ml_inter)
write.csv(lifestage_inter, file = paste0(data, "lifetage_discordance.csv"), row.names = F, quote = F)

```

TX_paternal_00029809 TX_paternal_00025458 TX_paternal_00008809

# examine AF
```{r}

```


# examine gene annotation
```{r}

```



# plot the phenotype ~ genotype

ml-mp mp-ml	3	A4:30274594 A3:13387496 A1:455264496
TX_paternal_00029809 TX_paternal_00025458 TX_paternal_00008809
```{r}
data <- "/Users/yuanmeng/Library/CloudStorage/OneDrive-UniversityofToronto/Manuscripts/rumex_eqtl/"
normalized_counts_mln <- read.table(paste0(data,"normalized_counts_mln.bed"), header = T)
normalized_counts_mpn <- read.table(paste0(data,"normalized_counts_mpn.bed"), header = T)

expr_l <- normalized_counts_mln %>% filter(gene_id == "TX_paternal_00029809")
expr_p <- normalized_counts_mpn %>% filter(gene_id == "TX_paternal_00029809")

geno <- read.table(paste0(data,"A4_30274594.vcf"), header = T)
geno <- as.data.frame(t(geno[,c(5:79)]))
geno <- geno %>% mutate(id = row.names(geno))
colnames(geno) <- c("genotype", "id")

expr_l <- as.data.frame(t(expr_l[,c(5:79)]))
expr_p <- as.data.frame(t(expr_p[,c(5:77)]))
expr_l <- expr_l %>% mutate(id = row.names(expr_l))
colnames(expr_l) <- c("leaf", "id")
expr_p <- expr_p %>% mutate(id = row.names(expr_p))
colnames(expr_p) <- c("pollen", "id")

eqtl_lp1 <- geno %>% left_join(expr_l, by = "id") %>% left_join(expr_p, by = "id")
eqtl_lp1$genotype <- as.factor(eqtl_lp1$genotype)
table(eqtl_lp1$genotype)
write.table(eqtl_lp1, file = paste0(data, "eqtl_lp1.txt"), quote = F)
eqtl_lp1 <- gather(eqtl_lp1, lifestage, Expression, leaf:pollen)

levels(eqtl_lp1$genotype) <- c("./.","GG", "GC","CC")
eqtl_lp1 <- eqtl_lp1 %>% filter(genotype != "./.")
write.table(eqtl_lp1, file = paste0(data, "eqtl_lp1_final.txt"), quote = F)

ggplot(eqtl_lp1, aes(x=genotype, y=Expression, colour=lifestage)) + 
  geom_boxplot(position=position_dodge2(0.85)) + 
  geom_jitter(shape=16, position=position_jitterdodge(0.2), alpha=0.3) + 
  theme_classic() + theme(text = element_text(size = 18))+ scale_colour_manual(values = c("orange", "purple"))+labs(x="Genotype", y="Expression")
```

ml-mp mp-ml	3	A4:30274594 A3:13387496 A1:455264496
TX_paternal_00029809 TX_paternal_00025458 TX_paternal_00008809
```{r}
expr_l <- normalized_counts_mln %>% filter(gene_id == "TX_paternal_00025458")
expr_p <- normalized_counts_mpn %>% filter(gene_id == "TX_paternal_00025458")

geno <- read.table(paste0(data,"A3_13387496.vcf"), header = T)
geno <- as.data.frame(t(geno[,c(5:79)]))
geno <- geno %>% mutate(id = row.names(geno))
colnames(geno) <- c("genotype", "id")

expr_l <- as.data.frame(t(expr_l[,c(5:79)]))
expr_p <- as.data.frame(t(expr_p[,c(5:77)]))
expr_l <- expr_l %>% mutate(id = row.names(expr_l))
colnames(expr_l) <- c("leaf", "id")
expr_p <- expr_p %>% mutate(id = row.names(expr_p))
colnames(expr_p) <- c("pollen", "id")

eqtl_lp2 <- geno %>% left_join(expr_l, by = "id") %>% left_join(expr_p, by = "id")
write.table(eqtl_lp2, file = paste0(data, "eqtl_lp2.txt"), quote = F)

eqtl_lp2$genotype <- as.factor(eqtl_lp2$genotype)
table(eqtl_lp2$genotype)
levels(eqtl_lp2$genotype) <- c("./.","GG", "GA")
eqtl_lp2 <- eqtl_lp2 %>% filter(genotype != "./.")

eqtl_lp2 <- gather(eqtl_lp2, lifestage, Expression, leaf:pollen)
write.table(eqtl_lp2, file = paste0(data, "eqtl_lp2_final.txt"), quote = F)

ggplot(eqtl_lp2, aes(x=genotype, y=Expression, colour=lifestage)) + 
  geom_boxplot(position=position_dodge2(0.85)) + 
  geom_jitter(shape=16, position=position_jitterdodge(0.2), alpha=0.3) + 
  theme_classic() + theme(text = element_text(size = 18))+ scale_colour_manual(values = c("orange", "purple"))+labs(x="Genotype", y="Expression")
```
ml-mp mp-ml	3	A4:30274594 A3:13387496 A1:455264496
TX_paternal_00029809 TX_paternal_00025458 TX_paternal_00008809
```{r}
expr_l <- normalized_counts_mln %>% filter(gene_id == "TX_paternal_00008809")
expr_p <- normalized_counts_mpn %>% filter(gene_id == "TX_paternal_00008809")
geno <- read.table(paste0(data,"A1_455264496.vcf"), header = T)
geno <- as.data.frame(t(geno[,c(5:79)]))
geno <- geno %>% mutate(id = row.names(geno))
colnames(geno) <- c("genotype", "id")
expr_l <- as.data.frame(t(expr_l[,c(5:79)]))
expr_p <- as.data.frame(t(expr_p[,c(5:77)]))
expr_l <- expr_l %>% mutate(id = row.names(expr_l))
colnames(expr_l) <- c("leaf", "id")
expr_p <- expr_p %>% mutate(id = row.names(expr_p))
colnames(expr_p) <- c("pollen", "id")

eqtl_lp3 <- geno %>% left_join(expr_l, by = "id") %>% left_join(expr_p, by = "id")
write.table(eqtl_lp3, file = paste0(data, "eqtl_lp3.txt"), quote = F)

eqtl_lp3$genotype <- as.factor(eqtl_lp3$genotype)
table(eqtl_lp3$genotype)
levels(eqtl_lp3$genotype) <- c("./.","AA", "AT")
eqtl_lp3 <- eqtl_lp3 %>% filter(genotype != "./.")
eqtl_lp3 <- gather(eqtl_lp3, lifestage, Expression, leaf:pollen)
write.table(eqtl_lp3, file = paste0(data, "eqtl_lp3_final.txt"), quote = F)

ggplot(eqtl_lp3, aes(x=genotype, y=Expression, colour=lifestage)) + 
  geom_boxplot(position=position_dodge2(0.85)) + 
  geom_jitter(shape=16, position=position_jitterdodge(0.2), alpha=0.3) + 
  theme_classic() + theme(text = element_text(size = 18))+ scale_colour_manual(values = c("orange", "purple"))+labs(x="Genotype", y="Expression")
```

```{r}
p1 <- ggplot(eqtl_lp1, aes(x=genotype, y=Expression, colour=lifestage)) + 
  geom_boxplot(position=position_dodge2(0.85)) + 
  geom_jitter(shape=16, position=position_jitterdodge(0.2), alpha=0.3) + 
  theme_classic() + theme(text = element_text(size = 12))+ scale_colour_manual(values = c("purple","orange"))+labs(x="Genotype", y="Expression")
p2 <- ggplot(eqtl_lp2, aes(x=genotype, y=Expression, colour=lifestage)) + 
  geom_boxplot(position=position_dodge2(0.85)) + 
  geom_jitter(shape=16, position=position_jitterdodge(0.2), alpha=0.3) + 
  theme_classic() + theme(text = element_text(size = 12))+ scale_colour_manual(values = c("purple","orange"))+labs(x="Genotype", y="Expression")
p3 <- ggplot(eqtl_lp3, aes(x=genotype, y=Expression, colour=lifestage)) + 
  geom_boxplot(position=position_dodge2(0.85)) + 
  geom_jitter(shape=16, position=position_jitterdodge(0.2), alpha=0.3) + 
  theme_classic() + theme(text = element_text(size = 12))+ scale_colour_manual(values = c("purple","orange"))+labs(x="Genotype", y="Expression")

p1 + p2 + p3 + plot_layout(guides = "collect")& theme(legend.position = "right")
```


# gene expression plot on autosomes
normalized_counts_mln
normalized_counts_mpn
```{r}
normalized_counts_mln <- normalized_counts_mln %>% mutate(l_mean = rowMeans(normalized_counts_mln[,c(5:79)]))
normalized_counts_mpn <- normalized_counts_mpn %>% mutate(p_mean = rowMeans(normalized_counts_mpn[,c(5:77)]))

mln <- normalized_counts_mln[,c(4,80)] #14804
mpn <- normalized_counts_mpn[,c(4,78)] # 13916

mlp <- inner_join(mln, mpn, by = "gene_id") # 12400

ggplot(pair1, aes(x = male_leaf, y = female_leaf, color = category)) + geom_point(size=0.8) + theme_classic() + labs(x="Male leaf", y="Female leaf") + theme(text = element_text(size = 11),axis.title.x = element_text(face="bold"))+ scale_colour_manual(values = c("grey", "black"))

ggplot(mlp, aes(x = l_mean, y =  p_mean)) + geom_point() + theme_classic()
```
