---
title: "expression phenotype files"
author: "Meng"
date: "2024-06-08"
output: html_document
---

```{r setup, include=FALSE}
library(ggplot2)
library(dplyr)
library("DESeq2")
library("stringr")
data <- "/Users/yuanmeng/Library/CloudStorage/OneDrive-UniversityofToronto/Manuscripts/rumex_eqtl/"
```

# male pollen and male leaf gene expression phenotype files

## remove unmatched samples and sex chroms before all nomalization
```{r}
readcnt <- read.table(paste0(data,"rhastTXeQTLrna_pollenleaf.txt"), header = T)
colname <- colnames(readcnt)
write.table(colname, file = paste0(data, "colname_pl.txt"), quote = F, row.names = F)

# simplify sample ids
colname <- read.table(paste0(data, "colname_pl.txt"))
colnames(readcnt) <- colname$V1

# remove 35aML (might be female), remove 24fM (low genomic coverage)
# remove RNA samples that don't have matching DNA samples
cols_to_remove <- c("24fMP", "35aMP", "35aML", "24fML", "26fML", "73aML")
readcnt <- readcnt %>%
  select(-all_of(cols_to_remove))	  

# remove genes with low read count 
readcnt$all_mean <- rowMeans(readcnt[,c(7:152)])
readcnt <- readcnt %>% filter(all_mean >= 5) # 20245

# remove sex chroms
gff <- read.table(paste0(data,"merged_TX_noMatPARlarge_txanno_genes.gtf"))
gff <- gff[,c(13,1,4,5)]
colnames(gff) <- c("gene","chrom", "start", "end")
colnames(readcnt)[1] <- "gene"
readcnt <- inner_join(gff, readcnt, by = "gene")
readcnt$chrom <- factor(readcnt$chrom)
levels(readcnt$chrom)
readcnt <- readcnt %>% filter(chrom != "X" & chrom != "Y") # 15681
write.table(readcnt, file = paste0(data,"readcnt_pollenLeaf_jun12.txt"), row.names = FALSE)

# prep for deseq2, need this for normalization by depth
readcnt_matrix <- as.matrix(readcnt[ , -c(1:9, 156)]) 
rownames(readcnt_matrix) <- readcnt[ , 1]
str(readcnt_matrix)
meta_data <- data.frame(colnames(readcnt_matrix))
colnames(meta_data) <- "id"
tissue <- rep(NA, length(meta_data$id))
tissue[grep("ML", meta_data$id)] <- "leaf"
tissue[grep("MP", meta_data$id)] <- "pollen"
meta_data <- data.frame(meta_data, tissue)
rownames(meta_data) <- meta_data[,1]
head(meta_data)
```

## run deseq2 to get nomalizaed read counts
```{r}
dds_tissue <- DESeqDataSetFromMatrix(countData = readcnt_matrix, 
                              colData = meta_data,
                              design = ~ tissue)
dds_tissue <- DESeq(dds_tissue)
resultsNames(dds_tissue) # "tissue_pollen_vs_leaf"
res_tissue <- results(dds_tissue)
head(res_tissue)
summary(res_tissue)
result_tissue <- res_tissue[complete.cases(res_tissue),]
nrow(res_tissue) # 15681
nrow(result_tissue) # 15681
```
out of 15681 with nonzero total read count
adjusted p-value < 0.1
LFC > 0 (up)       : 6076, 39%
LFC < 0 (down)     : 8751, 56%

### get nomalized read counts, separate MPR MLR 
```{r}
normalized_counts <- as.data.frame(counts(dds_tissue, normalized=T))
normalized_counts <- normalized_counts %>% mutate(gene=rownames(normalized_counts)) 
normalized_counts <- normalized_counts[,c(147,1:146)]
write.table(normalized_counts, file = paste0(data,"normalized_counts_pollenLeaf_jun12.txt"), row.names = FALSE)

col <- data.frame(colnames(normalized_counts))
normalized_counts_MP <- normalized_counts[,c(1,2:74)]
normalized_counts_ML <- normalized_counts[,c(1,75:147)]
```

### quantile normalizatin
```{r}
qn <- function(x){qqnorm(x, plot.it=FALSE)$x}
# quantile normalize, now columns genes, rows are inds
normalized_counts_MPn = apply(normalized_counts_MP[,-1], 1, qn)
normalized_counts_MPn <- as.data.frame(t(normalized_counts_MPn))
normalized_counts_MPn <- cbind(normalized_counts_MP[,1], normalized_counts_MPn)
colnames(normalized_counts_MPn) <- colnames(normalized_counts_MP)

normalized_counts_MLn = apply(normalized_counts_ML[,-1], 1, qn)
normalized_counts_MLn <- as.data.frame(t(normalized_counts_MLn))
normalized_counts_MLn <- cbind(normalized_counts_ML[,1], normalized_counts_MLn)
colnames(normalized_counts_MLn) <- colnames(normalized_counts_ML)

write.table(normalized_counts_MPn, file = paste0(data,"normalized_counts_MPn.txt"), row.names = FALSE, quote = F)

write.table(normalized_counts_MLn, file = paste0(data,"normalized_counts_MLn.txt"), row.names = FALSE, quote = F)
```



## get TSS info for each gene
### pollen
```{r}
gff_full <- read.table(paste0(data, "merged_TX_noMatPARlarge_txanno_gene_full.bed"))
colnames(gff_full) <- c("chrom", "start", "end", "strand")
gff_full <- inner_join(gff_full, gff, by = c("chrom", "start", "end"))

gff_full <- gff_full %>%
    mutate(start_tss = ifelse(strand == "+", start, end)) %>%
    mutate(end_tss = start_tss+1) 

gff_full <- gff_full[,c(1,6,7,5)]
colnames(gff_full) <- c("chrom", "start", "end", "gene")

normalized_counts_MPn <- inner_join(gff_full, normalized_counts_MPn, by = "gene")

# sort sample id
normalized_counts_MPn <- normalized_counts_MPn %>%
  select(1:4, sort(names(normalized_counts_MPn)[5:77]), everything())

#sort start pos for each chrom
normalized_counts_MPn <- normalized_counts_MPn %>%
  arrange(chrom, start) 

# remove extra letter in sample id (L or P)
col <- colnames(normalized_counts_MPn)
col <- str_replace(col, "MP$", "M")
colnames(normalized_counts_MPn) <- col

colnames(normalized_counts_MPn)[1] <- "#chr"
colnames(normalized_counts_MPn)[4] <- "gene_id"
write.table(normalized_counts_MPn, file = paste0(data,"normalized_counts_MPn.bed"), row.names = FALSE, quote = F, sep = "\t")
```

### leaf
```{r}
normalized_counts_MLn <- inner_join(gff_full, normalized_counts_MLn, by = "gene")

# sort sample id
normalized_counts_MLn <- normalized_counts_MLn %>%
  select(1:4, sort(names(normalized_counts_MLn)[5:77]), everything())

#sort start pos for each chrom
normalized_counts_MLn <- normalized_counts_MLn %>%
  arrange(chrom, start) 

# remove extra letter in sample id (L or P)
col <- colnames(normalized_counts_MLn)
col <- str_replace(col, "ML$", "M")
colnames(normalized_counts_MLn) <- col

colnames(normalized_counts_MLn)[1] <- "#chr"
colnames(normalized_counts_MLn)[4] <- "gene_id"
write.table(normalized_counts_MLn, file = paste0(data,"normalized_counts_MLn.bed"), row.names = FALSE, quote = F, sep = "\t")
```


## separate chroms at last
```{r}
colnames(normalized_counts_MPn)[1] <- "chrom"
colnames(normalized_counts_MLn)[1] <- "chrom"
normalized_counts_MPn_A4 <- normalized_counts_MPn %>% filter(chrom == "A4") 
normalized_counts_MLn_A4 <- normalized_counts_MLn %>% filter(chrom == "A4")

colnames(normalized_counts_MLn_A4)[1] <- "#chr"
colnames(normalized_counts_MPn_A4)[1] <- "#chr"
write.table(normalized_counts_MPn_A4, file = paste0(data,"normalized_counts_MPn_A4.bed"), row.names = FALSE, quote = F, sep = "\t")
write.table(normalized_counts_MLn_A4, file = paste0(data,"normalized_counts_MLn_A4.bed"), row.names = FALSE, quote = F, sep = "\t")

```

# combine pollen and leaf expression, add covariate
next steps
double the phenotype ids?
```{r}
qn <- function(x){qqnorm(x, plot.it=FALSE)$x}
# quantile normalize, now columns genes, rows are inds
normalized_counts_n = apply(normalized_counts_MP[,-1], 1, qn)
normalized_counts_n <- as.data.frame(t(normalized_counts_n))
normalized_counts_n <- cbind(normalized_counts_MP[,1], normalized_counts_n)
colnames(normalized_counts_n) <- colnames(normalized_counts_MP)

normalized_counts_n <- inner_join(gff_full, normalized_counts_n, by = "gene")

# sort sample id
normalized_counts_n <- normalized_counts_n %>%
  select(1:4, sort(names(normalized_counts_n)[5:77]), everything())

#sort start pos for each chrom
normalized_counts_n <- normalized_counts_n %>%
  arrange(chrom, start) 

# remove extra letter in sample id (L or P)
# need to think about this
col <- colnames(normalized_counts_n)
col <- str_replace(col, "ML$", "M")
colnames(normalized_counts_n) <- col

colnames(normalized_counts_n)[1] <- "#chr"
colnames(normalized_counts_n)[4] <- "gene_id"
write.table(normalized_counts_n, file = paste0(data,"normalized_counts_n.bed"), row.names = FALSE, quote = F, sep = "\t")
```



