---
title: "nucleotide diversity"
author: "Meng"
date: "2025-01-17"
output: html_document
---

```{r setup, include=FALSE}
library(ggplot2)
library(dplyr)
library(patchwork)
library(car)
data <- "/Users/yuanmeng/Library/CloudStorage/OneDrive-UniversityofToronto/Manuscripts/rumex_eqtl/popgen/"
data2 <- "/Users/yuanmeng/Library/CloudStorage/OneDrive-UniversityofToronto/Manuscripts/rumex_eqtl/eqtl/"
```


# pi and tajD data
```{r}
# pi <- read.table(paste0(data,"syn_pi.txt"), header = T) # 119485
# pi <- pi %>% filter(no_sites >= 50) # 20621
pi <- read.table(paste0(data,"syn.genewise_pi.txt"), header = T) # 27520
pi <- pi %>% filter(no_sites >= 50) # 22093

# tajd <- read.table(paste0(data,"syn.Tajima.D"), header = T) # 119392
# tajd <- tajd %>% filter(N_SNPS >= 10) # 3327, 11739
tajd <- read.csv(paste0(data,"syn.genewise_tajd.csv")) # 27519
tajd <- tajd[complete.cases(tajd),] # 7139
tajd <- tajd %>% filter(sites >= 20) # 7103

colnames(tajd)[1] <- "chrom"
colnames(pi)[2:4] <- c("chrom", "start", "end")
tajd$chrom <- as.character(tajd$chrom)
pi$chrom <- as.character(pi$chrom)


quantile(pi$avg_pi, 0.95) # 0.005546033
pi_top <- pi %>% filter(avg_pi > 0.005546033)
quantile(tajd$TajimaD, 0.95) # -1.40034
tajd_top <- tajd %>% filter(TajimaD > -1.40034)
```

```{r}
p1 <- ggplot(pi, aes(x = window_pos_1, y = avg_pi)) + geom_point(alpha = 0.3) + facet_grid(. ~ chrom, scales = "free_x", switch = "x", space = "free_x") + theme_classic()
p2 <- ggplot(tajd, aes(x = BIN_START, y = TajimaD)) + geom_point(alpha = 0.3) + facet_grid(. ~ chrom, scales = "free_x", switch = "x", space = "free_x") + theme_classic()

p1 / p2
ggsave(paste0(data,"pi_tajD.pdf"))
```


# Categories of eQTLs
## specific vs shared
### life-stage
```{r}
ML_egene <-read.table(paste0(data2, "ML_egene_eqtl.txt"))
MP_egene <-read.table(paste0(data2, "MP_egene_eqtl.txt"))
FL_egene <-read.table(paste0(data2, "FL_egene_eqtl.txt"))
colnames(ML_egene) <- "gene"
colnames(MP_egene) <- "gene"
colnames(FL_egene) <- "gene"

gff <- read.table("/Users/yuanmeng/Library/CloudStorage/OneDrive-UniversityofToronto/Manuscripts/rumex_eqtl/merged_TX_noMatPARlarge_txanno_genes.gtf")
gff <- gff[,c(13,1,4,5)]
colnames(gff) <- c("gene","chrom", "start", "end")
gff$chrom <- str_sub(gff$chrom, 2)

# life-stage specific vs shared
l_specific <- ML_egene %>% anti_join(MP_egene, by = "gene") # 1994
p_specific <- MP_egene %>% anti_join(ML_egene, by = "gene") # 1048
pl_shared <- inner_join(ML_egene, MP_egene, by = "gene") # 427

l_specific_pi <- l_specific %>% inner_join(gff, by = "gene") %>% 
    inner_join(pi, by = c("chrom", "start", "end")) # 1749
p_specific_pi <- p_specific %>% inner_join(gff, by = "gene") %>% 
    inner_join(pi, by = c("chrom", "start", "end")) # 925
pl_shared_pi <- pl_shared %>% inner_join(gff, by = "gene") %>% 
    inner_join(pi, by = c("chrom", "start", "end")) # 352

l_specific_pi$Category <- "Leaf-specific"
p_specific_pi$Category <- "Pollen-specific"
pl_shared_pi$Category <- "Shared"

pi_all <- rbind(l_specific_pi, p_specific_pi, pl_shared_pi)
leveneTest(data = pi_all, avg_pi ~ Category) # 0.364

group_by(pi_all, Category) %>%
  summarise(
    count = n(),
    mean = mean(avg_pi, na.rm = TRUE),
    sd = sd(avg_pi, na.rm = TRUE)
  )

aov <- aov(avg_pi ~ Category, data = pi_all)
summary(aov) # 0.701
TukeyHSD(aov)
```
Category count mean sd
Leaf-specific	1749	0.002143057	0.003051649	
Pollen-specific	925	0.002242397	0.004536414	
Shared	352	0.002280058	0.003380427

Df  Sum Sq   Mean Sq F value Pr(>F)
Category       2 0.00001 4.624e-06   0.356  0.701
Residuals   3023 0.03930 1.300e-05 

Tukey multiple comparisons of means
    95% family-wise confidence level

Fit: aov(formula = avg_pi ~ Category, data = pi_all)

$Category
                                      diff           lwr          upr     p adj
Pollen-specific-Leaf-specific 9.934031e-05 -0.0002444019 0.0004430825 0.7765345
Shared-Leaf-specific          1.370010e-04 -0.0003569282 0.0006309301 0.7921597
Shared-Pollen-specific        3.766065e-05 -0.0004918457 0.0005671670 0.9847832

```{r}
gff <- read.table("/Users/yuanmeng/Library/CloudStorage/OneDrive-UniversityofToronto/Manuscripts/rumex_eqtl/merged_TX_noMatPARlarge_txanno_genes.gtf")
gff <- gff[,c(13,1,4,5)]
colnames(gff) <- c("gene","chrom", "start", "end")
gff$chrom <- str_sub(gff$chrom, 2)

l_specific_tajd <- l_specific %>% inner_join(gff, by = "gene") %>% 
    inner_join(tajd, by = c("chrom", "start", "end")) # 655
p_specific_tajd <- p_specific %>% inner_join(gff, by = "gene") %>% 
    inner_join(tajd, by = c("chrom", "start", "end")) # 340
pl_shared_tajd <- pl_shared %>% inner_join(gff, by = "gene") %>% 
    inner_join(tajd, by = c("chrom", "start", "end")) # 119

l_specific_tajd$Category <- "Leaf-specific"
p_specific_tajd$Category <- "Pollen-specific"
pl_shared_tajd$Category <- "Shared"

tajd_all <- rbind(l_specific_tajd, p_specific_tajd, pl_shared_tajd)
leveneTest(data = tajd_all, TajD_pop ~ Category) # p = 0.3527

group_by(tajd_all, Category) %>%
  summarise(
    count = n(),
    mean = mean(TajD_pop, na.rm = TRUE),
    sd = sd(TajD_pop, na.rm = TRUE)
  )

aov <- aov(TajD_pop ~ Category, data = tajd_all)
summary(aov) # p = 0.769
TukeyHSD(aov)
```
Levene's Test for Homogeneity of Variance (center = median)
        Df F value Pr(>F)
group    2  1.0431 0.3527
      1111    
      
Category count mean sd
Leaf-specific	655	-1.183879	0.3186621	
Pollen-specific	340	-1.172744	0.3029627	
Shared	119	-1.195571	0.3571017	      
      
Df Sum Sq Mean Sq F value Pr(>F)
Category       2   0.05 0.02664   0.263  0.769
Residuals   1111 112.57 0.10133 

Tukey multiple comparisons of means
    95% family-wise confidence level

Fit: aov(formula = TajD_pop ~ Category, data = tajd_all)

$Category
                                     diff         lwr        upr     p adj
Pollen-specific-Leaf-specific  0.01113451 -0.03879979 0.06106880 0.8599554
Shared-Leaf-specific          -0.01169280 -0.08613580 0.06275019 0.9278465
Shared-Pollen-specific        -0.02282731 -0.10239576 0.05674114 0.7790886

```{r}
p1 <- ggplot(pi_all, aes(x = Category, y = avg_pi)) + geom_boxplot()
p2 <- ggplot(tajd_all, aes(x = Category, y = TajD_pop)) + geom_boxplot()
p1+p2
ggsave(paste0(data,"lifestage_pi_tajd.pdf"), width = 7, height = 3, units = "in")
```

### sex
```{r}
# sex specific vs shared
f_specific <- FL_egene %>% anti_join(ML_egene, by = "gene") # 2072
m_specific <- ML_egene %>% anti_join(FL_egene, by = "gene") # 1051
sex_shared <- inner_join(ML_egene, FL_egene, by = "gene") # 1370

f_specific_pi <- f_specific %>% inner_join(gff, by = "gene") %>% 
    inner_join(pi, by = c("chrom", "start", "end")) # 1889
m_specific_pi <- m_specific %>% inner_join(gff, by = "gene") %>% 
    inner_join(pi, by = c("chrom", "start", "end")) # 957
sex_shared_pi <- sex_shared %>% inner_join(gff, by = "gene") %>% 
    inner_join(pi, by = c("chrom", "start", "end")) # 1144

m_specific_pi$Category <- "Male-specific"
f_specific_pi$Category <- "Female-specific"
sex_shared_pi$Category <- "Shared"
sex_pi <- rbind(m_specific_pi, f_specific_pi, sex_shared_pi)

group_by(sex_pi, Category) %>%
  summarise(
    count = n(),
    mean = mean(avg_pi, na.rm = TRUE),
    sd = sd(avg_pi, na.rm = TRUE)
  )

leveneTest(data = sex_pi, avg_pi ~ Category) # p = 7.706e-06
aov <- aov(avg_pi ~ Category, data = sex_pi)
summary(aov) # p = 0.000257
TukeyHSD(aov)
```
Category count mean sd 
Female-specific	1889	0.001881269	0.002545503	
Male-specific	957	0.001990252	0.002605078	
Shared	1144	0.002313038	0.003468735	

              Df  Sum Sq   Mean Sq F value Pr(>F)
Category       2 0.00001 4.624e-06   0.356  0.701
Residuals   3023 0.03930 1.300e-05 

Df  Sum Sq   Mean Sq F value   Pr(>F)    
Category       2 0.00013 6.747e-05   8.284 0.000257 ***
Residuals   3987 0.03247 8.140e-06                     
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Tukey multiple comparisons of means
    95% family-wise confidence level

Fit: aov(formula = avg_pi ~ Category, data = sex_pi)

$Category
                                      diff           lwr          upr     p adj
Male-specific-Female-specific 0.0001089830 -1.565112e-04 0.0003744771 0.6007208
Shared-Female-specific        0.0004317691  1.810909e-04 0.0006824474 0.0001622
Shared-Male-specific          0.0003227862  2.966075e-05 0.0006159116 0.0266630

```{r}
f_specific_tajd <- f_specific %>% inner_join(gff, by = "gene") %>% 
    inner_join(tajd, by = c("chrom", "start", "end")) # 690
m_specific_tajd <- m_specific %>% inner_join(gff, by = "gene") %>% 
    inner_join(tajd, by = c("chrom", "start", "end")) # 360
sex_shared_tajd <- sex_shared %>% inner_join(gff, by = "gene") %>% 
    inner_join(tajd, by = c("chrom", "start", "end")) # 414

m_specific_tajd$Category <- "Male-specific"
f_specific_tajd$Category <- "Female-specific"
sex_shared_tajd$Category <- "Shared"
sex_tajd <- rbind(m_specific_tajd, f_specific_tajd, sex_shared_tajd)

group_by(sex_tajd, Category) %>%
  summarise(
    count = n(),
    mean = mean(TajD_pop, na.rm = TRUE),
    sd = sd(TajD_pop, na.rm = TRUE)
  )

leveneTest(data = sex_tajd, TajD_pop ~ Category) # p = 0.7933
aov <- aov(TajD_pop ~ Category, data = sex_tajd)
summary(aov) # p = 0.799
TukeyHSD(aov)
```

Category count mean sd 
Female-specific	690	-1.174386	0.3214620	
Male-specific	360	-1.185023	0.3214939	
Shared	414	-1.186245	0.3277562	

              Df Sum Sq Mean Sq F value Pr(>F)
Category       2   0.05 0.02339   0.224  0.799
Residuals   1461 152.67 0.10450 

Tukey multiple comparisons of means
    95% family-wise confidence level

Fit: aov(formula = TajD_pop ~ Category, data = sex_tajd)

$Category
                                      diff         lwr        upr     p adj
Male-specific-Female-specific -0.010636268 -0.05994441 0.03867187 0.8683912
Shared-Female-specific        -0.011858696 -0.05900630 0.03528890 0.8254600
Shared-Male-specific          -0.001222428 -0.05587604 0.05343118 0.9984830
      
box plot of pi and tajD for 3 categories
```{r}
p1 <- ggplot(sex_pi, aes(x = Category, y = avg_pi)) + geom_boxplot()
p2 <- ggplot(sex_tajd, aes(x = Category, y = TajD_pop)) + geom_boxplot()
p1 + p2
ggsave(paste0(data,"sex_pi_tajd.pdf"), width = 7, height = 3, units = "in")
```


## concordant vs discordant between life stages
```{r}
# concordant
pl_concordant <- ml_mp %>% filter(slope.x * slope.y > 0) # 1621
pl_discordant <- ml_mp %>% filter(slope.x * slope.y < 0) # 75

# 
# sex_discordant <- ml_fl %>% filter(slope.x * slope.y < 0) # 0
# discordant between sexes
# dis_sex <- read.csv(paste0(data, "dis_sex.csv")) # 170
# discordant between life-stages
# dis_pl <- read.csv(paste0(data, "dis_pl.csv")) # 974
```

check concordant for pi and tajD outliers
```{r}
sex_concordant <- ml_fl %>% filter(slope.x * slope.y > 0) # 5516

# m_specific_pi <- m_specific %>% inner_join(pi, by = "chrom") %>%
#     filter(pos >= window_pos_1 & pos <= window_pos_2 ) # 5180

```


## eQTLs for Genotype × Sex interaction (5 eqtls)
```{r}
pval_adj_bh <- read.csv(paste0(data2, "sex_interaction_padj_bh01.csv")) # 5
pval_adj_bh <- pval_adj_bh[, 1:2]
colnames(pval_adj_bh)[1] <- "gene"
pval_adj_bh <- pval_adj_bh %>% inner_join(gff, by = "gene") %>% 
    left_join(pi, by = c("chrom", "start", "end")) %>%
    left_join(tajd, by = c("chrom", "start", "end")) 
write.table(pval_adj_bh, file = paste0(data, "pval_adj_bh_pi.txt"), quote = F, row.names = F)
```
