---
title: "nucleotide diversity"
author: "Meng"
date: "2025-01-17"
output: html_document
---

```{r setup, include=FALSE}
library(ggplot2)
library(dplyr)
library(patchwork)
library(car)
library(stringr)
data <- "/Users/yuanmeng/Library/CloudStorage/OneDrive-UniversityofToronto/Manuscripts/rumex_eqtl/popgen/"
data2 <- "/Users/yuanmeng/Library/CloudStorage/OneDrive-UniversityofToronto/Manuscripts/rumex_eqtl/eqtl/"
```


# pi and tajD data
```{r}
pi <- read.table(paste0(data,"syn.genewise_pi.txt"), header = T) # 27520
pi <- pi[complete.cases(pi),] # 26690
pi <- pi %>% filter(no_sites >= 50) # 22093

tajd <- read.table(paste0(data,"syn.genewise_tajima_d.txt"), header = T) # 27520
tajd <- tajd[complete.cases(tajd),] # 23313
tajd <- tajd %>% filter(no_sites >= 50) # 20621

colnames(pi)[2:4] <- c("chrom", "start", "end")
colnames(tajd)[2:4] <- c("chrom", "start", "end")
tajd$chrom <- as.character(tajd$chrom)
pi$chrom <- as.character(pi$chrom)
```

```{r}
p1 <- ggplot(pi, aes(x = start, y = avg_pi)) + geom_point(alpha = 0.3) + facet_grid(. ~ chrom, scales = "free_x", switch = "x", space = "free_x") + theme_classic()
p2 <- ggplot(tajd, aes(x = start, y = tajima_d)) + geom_point(alpha = 0.3) + facet_grid(. ~ chrom, scales = "free_x", switch = "x", space = "free_x") + theme_classic()

p1 / p2
ggsave(paste0(data,"pi_tajD.pdf"))
```


# Categories of eQTLs
```{r}
ML_eqtl_multigene <- read.csv(paste0(data2, "ML_eqtl_multigene.csv"))
FL_eqtl_multigene <- read.csv(paste0(data2, "FL_eqtl_multigene.csv"))
MP_eqtl_multigene <- read.csv(paste0(data2, "MP_eqtl_multigene.csv"))
ML_eqtl_sig <-read.csv(paste0(data2, "ML_eqtl_sig_Nov.csv"))
FL_eqtl_sig <-read.csv(paste0(data2, "FL_eqtl_sig_Nov.csv"))
MP_eqtl_sig <-read.csv(paste0(data2, "MP_eqtl_sig_Nov.csv"))

ML_egene <- ML_eqtl_sig %>% 
    anti_join(ML_eqtl_multigene, by = c("variant_id","maf")) %>% 
    dplyr::select(phenotype_id) %>% distinct() # 2359
FL_egene <- FL_eqtl_sig %>% 
    anti_join(FL_eqtl_multigene, by = c("variant_id","maf")) %>% 
    dplyr::select(phenotype_id) %>% distinct() # 3351
MP_egene <- MP_eqtl_sig %>% 
    anti_join(MP_eqtl_multigene, by = c("variant_id","maf")) %>% 
    dplyr::select(phenotype_id) %>% distinct() # 1428

colnames(ML_egene) <- "gene"
colnames(MP_egene) <- "gene"
colnames(FL_egene) <- "gene"

gff <- read.table("/Users/yuanmeng/Library/CloudStorage/OneDrive-UniversityofToronto/Manuscripts/rumex_eqtl/merged_TX_noMatPARlarge_txanno_genes.gtf")
gff <- gff[,c(13,1,4,5)]
colnames(gff) <- c("gene","chrom", "start", "end")
gff$chrom <- str_sub(gff$chrom, 2)
```

## specific vs shared
### life-stage
```{r}
# life-stage specific vs shared
l_specific <- ML_egene %>% anti_join(MP_egene, by = "gene") # 1958
p_specific <- MP_egene %>% anti_join(ML_egene, by = "gene") # 1027
pl_shared <- inner_join(ML_egene, MP_egene, by = "gene") # 401

l_specific_pi <- l_specific %>% inner_join(gff, by = "gene") %>% 
    inner_join(pi, by = c("chrom", "start", "end")) # 1720
p_specific_pi <- p_specific %>% inner_join(gff, by = "gene") %>% 
    inner_join(pi, by = c("chrom", "start", "end")) # 906
pl_shared_pi <- pl_shared %>% inner_join(gff, by = "gene") %>% 
    inner_join(pi, by = c("chrom", "start", "end")) # 331

l_specific_pi$Category <- "Leaf-specific"
p_specific_pi$Category <- "Pollen-specific"
pl_shared_pi$Category <- "Shared"

pi_all <- rbind(l_specific_pi, p_specific_pi, pl_shared_pi)
leveneTest(data = pi_all, avg_pi ~ Category) # p = 0.3329

group_by(pi_all, Category) %>%
  summarise(
    count = n(),
    mean = mean(avg_pi, na.rm = TRUE),
    sd = sd(avg_pi, na.rm = TRUE)
  )

aov <- aov(avg_pi ~ Category, data = pi_all)
summary(aov) # p = 0.611
TukeyHSD(aov)
```
Category count mean sd
Leaf-specific	1720	0.002151170	0.003071373	
Pollen-specific	906	0.002263338	0.004579506	
Shared	331	0.002327930	0.003458641

Df  Sum Sq   Mean Sq F value Pr(>F)
Category       2 0.00001 6.535e-06   0.493  0.611
Residuals   2954 0.03914 1.325e-05


```{r}
l_specific_tajd <- l_specific %>% inner_join(gff, by = "gene") %>% 
    inner_join(tajd, by = c("chrom", "start", "end")) # 1697
p_specific_tajd <- p_specific %>% inner_join(gff, by = "gene") %>% 
    inner_join(tajd, by = c("chrom", "start", "end")) # 898
pl_shared_tajd <- pl_shared %>% inner_join(gff, by = "gene") %>% 
    inner_join(tajd, by = c("chrom", "start", "end")) # 325

l_specific_tajd$Category <- "Leaf-specific"
p_specific_tajd$Category <- "Pollen-specific"
pl_shared_tajd$Category <- "Shared"

tajd_all <- rbind(l_specific_tajd, p_specific_tajd, pl_shared_tajd)
leveneTest(data = tajd_all, tajima_d ~ Category) # p = 0.1651

group_by(tajd_all, Category) %>%
  summarise(
    count = n(),
    mean = mean(tajima_d, na.rm = TRUE),
    sd = sd(tajima_d, na.rm = TRUE)
  )

aov <- aov(tajima_d ~ Category, data = tajd_all)
summary(aov) # p = 0.157
TukeyHSD(aov)
```
  
Category count mean sd
Leaf-specific	1697	-0.9125938	0.2232674	
Pollen-specific	898	-0.9157521	0.2076022	
Shared	325	-0.8888459	0.2585612		

Df Sum Sq Mean Sq F value Pr(>F)
Category       2   0.18 0.09204   1.853  0.157
Residuals   2917 144.86 0.04966 


```{r}
p1 <- ggplot(pi_all, aes(x = Category, y = avg_pi)) + geom_boxplot()
p2 <- ggplot(tajd_all, aes(x = Category, y = tajima_d)) + geom_boxplot()
p1+p2
ggsave(paste0(data,"lifestage_pi_tajd.pdf"), width = 7, height = 3, units = "in")
```

### sex
```{r}
# sex specific vs shared
f_specific <- FL_egene %>% anti_join(ML_egene, by = "gene") # 2037
m_specific <- ML_egene %>% anti_join(FL_egene, by = "gene") # 1045
sex_shared <- inner_join(ML_egene, FL_egene, by = "gene") # 1314

f_specific_pi <- f_specific %>% inner_join(gff, by = "gene") %>% 
    inner_join(pi, by = c("chrom", "start", "end")) # 1856
m_specific_pi <- m_specific %>% inner_join(gff, by = "gene") %>% 
    inner_join(pi, by = c("chrom", "start", "end")) # 950
sex_shared_pi <- sex_shared %>% inner_join(gff, by = "gene") %>% 
    inner_join(pi, by = c("chrom", "start", "end")) # 1101

m_specific_pi$Category <- "Male-specific"
f_specific_pi$Category <- "Female-specific"
sex_shared_pi$Category <- "Shared"
sex_pi <- rbind(m_specific_pi, f_specific_pi, sex_shared_pi)

group_by(sex_pi, Category) %>%
  summarise(
    count = n(),
    mean = mean(avg_pi, na.rm = TRUE),
    sd = sd(avg_pi, na.rm = TRUE)
  )

leveneTest(data = sex_pi, avg_pi ~ Category) # p = 3.938e-06
# use non parametric test
kruskal.test(avg_pi ~ Category, data = sex_pi) # p-value = 0.5383
pairwise.wilcox.test(sex_pi$avg_pi, sex_pi$Category, p.adjust.method = "BH")
```
Category count mean sd 
Female-specific	1856	0.001883272	0.002549293	
Male-specific	950	0.001994788	0.002613055	
Shared	1101	0.002339245	0.003520300

Kruskal-Wallis rank sum test

data:  avg_pi by Category
Kruskal-Wallis chi-squared = 1.2385, df = 2, p-value = 0.5383


```{r}
f_specific_tajd <- f_specific %>% inner_join(gff, by = "gene") %>% 
    inner_join(tajd, by = c("chrom", "start", "end")) # 1833
m_specific_tajd <- m_specific %>% inner_join(gff, by = "gene") %>% 
    inner_join(tajd, by = c("chrom", "start", "end")) # 938
sex_shared_tajd <- sex_shared %>% inner_join(gff, by = "gene") %>% 
    inner_join(tajd, by = c("chrom", "start", "end")) # 1084

m_specific_tajd$Category <- "Male-specific"
f_specific_tajd$Category <- "Female-specific"
sex_shared_tajd$Category <- "Shared"
sex_tajd <- rbind(m_specific_tajd, f_specific_tajd, sex_shared_tajd)

group_by(sex_tajd, Category) %>%
  summarise(
    count = n(),
    mean = mean(tajima_d, na.rm = TRUE),
    sd = sd(tajima_d, na.rm = TRUE)
  )

leveneTest(data = sex_tajd, tajima_d ~ Category) # p = 0.0004625
# use non parametric test
kruskal.test(tajima_d ~ Category, data = sex_tajd) # p-value = 4.523e-05
pairwise.wilcox.test(sex_tajd$tajima_d, sex_tajd$Category, p.adjust.method = "BH")
```

Category count mean sd 
Female-specific	1833	-0.9341454	0.1970029	
Male-specific	938	-0.9288674	0.2009582	
Shared	1084	-0.8913921	0.2502280	

Kruskal-Wallis rank sum test

data:  tajima_d by Category
Kruskal-Wallis chi-squared = 20.008, df = 2, p-value = 4.523e-05

Pairwise comparisons using Wilcoxon rank sum test with continuity correction 

data:  sex_tajd\$tajima_d and sex_tajd$Category 

Female-specific Male-specific
Male-specific 0.85486         -            
Shared        0.00011         0.00033      

P value adjustment method: BH 


box plot of pi and tajD for 3 categories
```{r}
p1 <- ggplot(sex_pi, aes(x = Category, y = avg_pi)) + geom_boxplot()
p2 <- ggplot(sex_tajd, aes(x = Category, y = tajima_d)) + geom_boxplot()
p1 + p2
ggsave(paste0(data,"sex_pi_tajd.pdf"), width = 7, height = 3, units = "in")
```


## eGenes in ML/MP vs differential expression between ML & MP
```{r}
egene_all <- full_join(ML_egene, MP_egene, by = "gene") # 3386
pl_eqtl_sig <- read.csv(paste0(data2, "pl_eqtl_sig.csv"))
pl_egene <- as.data.frame(unique(pl_eqtl_sig$phenotype_id)) # 1564
colnames(pl_egene) <- "gene"

pl_not <- anti_join(egene_all, pl_egene, by = "gene") # 2063

pl_egene_pi <- pl_egene %>% inner_join(gff, by = "gene") %>% 
    inner_join(pi, by = c("chrom", "start", "end")) # 1306
pl_egene_tajd <- pl_egene %>% inner_join(gff, by = "gene") %>% 
    inner_join(tajd, by = c("chrom", "start", "end")) # 1286

pl_not_pi <- pl_not %>% inner_join(gff, by = "gene") %>% 
    inner_join(pi, by = c("chrom", "start", "end")) # 1870
pl_not_tajd <- pl_not %>% inner_join(gff, by = "gene") %>% 
    inner_join(tajd, by = c("chrom", "start", "end")) # 1847

pl_egene_pi$Category <- "DE"
pl_not_pi$Category <- "noDE"
pl_egene_tajd$Category <- "DE"
pl_not_tajd$Category <- "noDE"

pl_pi <- rbind(pl_egene_pi, pl_not_pi)    
pl_tajd <- rbind(pl_egene_tajd, pl_not_tajd)    

group_by(pl_tajd, Category) %>%
  summarise(
    count = n(),
    mean = mean(tajima_d, na.rm = TRUE),
    sd = sd(tajima_d, na.rm = TRUE)
  )

leveneTest(data = pl_tajd, tajima_d ~ Category) # p = 0.1584
t.test(tajima_d ~ Category, data = pl_tajd, var.equal = T) # p = 0.0005815

group_by(pl_pi, Category) %>%
  summarise(
    count = n(),
    mean = mean(avg_pi, na.rm = TRUE),
    sd = sd(avg_pi, na.rm = TRUE)
  )

leveneTest(data = pl_pi, avg_pi ~ Category) # p = 4.687e-06 ***
t.test(avg_pi ~ Category, data = pl_pi, var.equal = F) # p = 0.0001305

```
tajd
Category count mean sd 
DE	1286	-0.8961768	0.2323016	
noDE	1847	-0.9235857	0.2095142

Two Sample t-test

data:  tajima_d by Category
t = 3.4436, df = 3131, p-value = 0.0005815
alternative hypothesis: true difference in means between group DE and group noDE is not equal to 0
95 percent confidence interval:
 0.01180292 0.04301483
sample estimates:
  mean in group DE mean in group noDE 
        -0.8961768         -0.9235857 
         
pi 
Category count mean sd 
DE	1306	0.002499847	0.004570933	
noDE	1870	0.001961369	0.002642290	

Welch Two Sample t-test

data:  avg_pi by Category
t = 3.8334, df = 1912.5, p-value = 0.0001305
alternative hypothesis: true difference in means between group DE and group noDE is not equal to 0
95 percent confidence interval:
 0.0002629898 0.0008139671
sample estimates:
  mean in group DE mean in group noDE 
       0.002499847        0.001961369 

## eQTLs for Genotype × Sex interaction (5 eqtls)
```{r}
pval_adj_bh <- read.csv(paste0(data2, "sex_interaction_padj_bh01.csv")) # 5
pval_adj_bh <- pval_adj_bh[, 1:2]
colnames(pval_adj_bh)[1] <- "gene"
pval_adj_bh <- pval_adj_bh %>% inner_join(gff, by = "gene") %>% 
    left_join(pi, by = c("chrom", "start", "end")) %>%
    left_join(tajd, by = c("chrom", "start", "end")) 
write.table(pval_adj_bh, file = paste0(data, "pval_adj_bh_pi.txt"), quote = F, row.names = F)
```
