---
title: "expression phenotype files"
author: "Meng"
date: "2024-06-13"
output: html_document
---

```{r setup, include=FALSE}
library(ggplot2)
library(dplyr)
library("DESeq2")
library("stringr")
data <- "/Users/yuanmeng/Library/CloudStorage/OneDrive-UniversityofToronto/Manuscripts/rumex_eqtl/"
```


# female leaf gene expression phenotype files
phenotype_mf.Rmd: remove unmatched samples
```{r}
# unwanted samples already removed
readcnt_leaf <- read.table(paste0(data,"readcnt_leaf_sept.txt"), header = T, check.names = F)
col<-data.frame(colnames(readcnt_leaf))
readcnt_leaf <- readcnt_leaf %>%  mutate(m_mean = rowMeans(readcnt_leaf[,c(2:70)] )) %>% mutate(f_mean = rowMeans(readcnt_leaf[,c(71:144)] ))
# remove genes with low read counts
readcnt_leaf <- readcnt_leaf %>% filter(m_mean >= 5 | f_mean >= 5) # 37659 -> 19494

gff <- read.table(paste0(data,"merged_TX_noMatPARlarge_txanno_genes.gtf"))
gff <- gff[,c(13,1,4,5)]
colnames(gff) <- c("gene","chrom", "start", "end")
gff <- gff %>% mutate(chrom_new = case_when(
    chrom == "A1" ~ "A1",
    chrom == "A2" ~ "A2",
    chrom == "A3" ~ "A3",
    chrom == "A4" ~ "A4",
    chrom == "X" ~ "X",
    chrom == "Y" & start < 45000000 ~ "PAR",
    chrom == "Y" & start >= 45000000 ~ "Y"
))
write.table(gff, file=paste0(data,"gff.txt"), row.names = FALSE, quote = F, sep = "\t")
gff <- read.table(paste0(data, "gff.txt"), header = T)
readcnt_leaf <- inner_join(gff, readcnt_leaf, by = "gene")

# keep only autosomes and PAR, figure out normalization on sex chroms later
readcnt_leaf <- readcnt_leaf %>% filter(chrom_new != "X" & chrom_new != "Y") # 19486 -> 16131
table(readcnt_leaf$chrom_new)
```
A1   A2   A3   A4  PAR 
5064 4778 2751 2399 1139

## perform normalization combined 
run deseq2 to get normalized read counts

```{r}
# prep for deseq2
# exclude gene info columns and all_mean column
readcnt_matrix<-as.matrix(readcnt_leaf[ , -c(1:5,149,150)]) 
rownames(readcnt_matrix) <- readcnt_leaf[ , 1]
str(readcnt_matrix)
meta_data <- data.frame(colnames(readcnt_matrix))

colnames(meta_data) <- "id"
sex <- rep(NA, length(meta_data$id))
sex[grep("M", meta_data$id)] <- "male"
sex[grep("F", meta_data$id)] <- "female"
meta_data <- data.frame(meta_data, sex)
rownames(meta_data) <- meta_data[,1]
head(meta_data)
write.table(meta_data, file=paste0(data,"meta_data_sex.txt"), row.names = FALSE, quote = F)

# run deseq2
dds <- DESeqDataSetFromMatrix(countData = readcnt_matrix, 
                              colData = meta_data,
                              design = ~ sex)
dds <- DESeq(dds)
resultsNames(dds) # "sex_male_vs_female"
res <- results(dds)
head(res)
summary(res)
result <- res[complete.cases(res),]
nrow(res) #   16131
nrow(result) # 16131

normalized_countsl <- as.data.frame(counts(dds, normalized=T))
normalized_countsl <- normalized_countsl %>% mutate(gene=rownames(normalized_countsl)) 
normalized_countsl <- normalized_countsl[,c(144,1:143)]
write.table(normalized_countsl, file = paste0(data,"normalized_countsl_sept.txt"), row.names = FALSE, quote = F)

# DE_sex<-as.data.frame(result)
# DE_sex <- DE_sex %>% mutate(gene = rownames(DE_sex))
# DE_sex <- DE_sex[,c(7,1:6)]
# DE_sex <- inner_join(DE_sex, gff, by = "gene")
# write.table(DE_sex, file = paste0(data,"DE_sex_A_PAR.txt"), row.names = FALSE, quote = F)
```

## quick check on expression PCA of autosomes and PAR
```{r}
vsdata <- vst(dds, blind=FALSE)
#+ geom_text(aes(label=name),vjust=2)
plotPCA(vsdata, intgroup=c("sex"))  
p1 <- plotPCA(vsdata, intgroup=c("sex")) 
#plotPCA(vsdata, intgroup=c("sex"))  + geom_text(aes(label=name),vjust=2)
```


## quantile normalizatin
```{r}
qn <- function(x){qqnorm(x, plot.it=FALSE)$x}

# quantile normalize, now columns genes, rows are inds
# combined
normalized_counts_ln = apply(normalized_countsl[,-1], 1, qn)
normalized_counts_ln <- as.data.frame(t(normalized_counts_ln))
normalized_counts_ln <- cbind(normalized_countsl[,1], normalized_counts_ln)
colnames(normalized_counts_ln) <- colnames(normalized_countsl)

```


## get TSS info for each gene
```{r}
gff_full <- read.table(paste0(data, "merged_TX_noMatPARlarge_txanno_gene_full.bed"))
colnames(gff_full) <- c("chrom", "start", "end", "strand")
gff_full <- inner_join(gff_full, gff, by = c("chrom", "start", "end"))
gff_full <- gff_full %>%
    mutate(start_tss = ifelse(strand == "+", start, end)) %>%
    mutate(end_tss = start_tss+1) 
gff_full <- gff_full[,c(6,7,8,5)]
colnames(gff_full) <- c("chrom", "start", "end", "gene")
write.table(gff_full, file = paste0(data, "gff_full.txt"), row.names = FALSE, quote = F, sep = "\t")
gff_full <- read.table(paste0(data, "gff_full.txt"), header = T)

# combined
normalized_counts_ln <- inner_join(gff_full, normalized_counts_ln, by = "gene")
#sort start pos for each chrom
normalized_counts_ln <- normalized_counts_ln %>%
  arrange(chrom, start) 

colnames(normalized_counts_ln)[4] <- "gene_id"
normalized_counts_ln$chrom <- factor(normalized_counts_ln$chrom)
table(normalized_counts_ln$chrom)

normalized_counts_ln_auto <- normalized_counts_ln %>% filter(chrom != "PAR") # 14992
normalized_counts_ln_PAR <- normalized_counts_ln %>% filter(chrom == "PAR") # 1139

colnames(normalized_counts_ln_PAR)[1] <- "#chr"
colnames(normalized_counts_ln_auto)[1] <- "#chr"

write.table(normalized_counts_ln, file = paste0(data,"normalized_counts_ln_full.bed"), row.names = FALSE, quote = F, sep = "\t")
write.table(normalized_counts_ln_PAR, file = paste0(data,"normalized_counts_ln_PAR.bed"), row.names = FALSE, quote = F, sep = "\t")
write.table(normalized_counts_ln_auto, file = paste0(data,"normalized_counts_ln_auto.bed"), row.names = FALSE, quote = F, sep = "\t")
```
A1   A2   A3   A4  PAR 
5061 4778 2748 2398 1139 

re-order the columns so it match the genotype data
```{r}
normalized_counts_ln <- read.table(paste0(data,"normalized_counts_ln_full.bed"), header = T, check.names = F)
new_order <- scan(paste0(data,"sex_143.txt"), what = character(), sep = "\n")
normalized_counts_ln <- normalized_counts_ln[, c(names(normalized_counts_ln)[1:4], new_order)]

normalized_counts_ln_auto <- normalized_counts_ln %>% filter(chrom != "PAR") # 14992
normalized_counts_ln_PAR <- normalized_counts_ln %>% filter(chrom == "PAR") # 1139

```

