---
title: "phenotype for female leaf and male leaf separate normalization"
author: "Meng"
date: "2024-09-23"
output: html_document
---

```{r setup, include=FALSE}
library(ggplot2)
library(dplyr)
library("DESeq2")
library("stringr")
data <- "/Users/yuanmeng/Library/CloudStorage/OneDrive-UniversityofToronto/Manuscripts/rumex_eqtl/"
```

## prep read cnt file
```{r}
readcnt_leaf <- read.table(paste0(data,"rhastTXeQTLrna.txt"), header = T)
head(readcnt_leaf)
colname <- colnames(readcnt_leaf)
#write.table(colname, file = paste0(data, "colname_mf.txt"), quote = F, row.names = F)
colname <- read.table(paste0(data, "colname_mf.txt"))
colnames(readcnt_leaf) <- colname$V1
# sort sample id, males then females
readcnt_leaf <- readcnt_leaf %>%
  dplyr::select(1:6, sort(names(readcnt_leaf)[7:161]), everything())
readcnt_leaf_m <- readcnt_leaf %>% dplyr::select(ends_with("M")) # 77
readcnt_leaf_f <- readcnt_leaf %>% dplyr::select(ends_with("F")) # 78

# remove 35aMLR (might be female), 40aF (might be male)
# remove 24fM (low genomic coverage)
# remove RNA samples that don't have matching DNA samples
# 26fML and 73aML dont have matching pollen sample but can be kept here
cols_to_remove <- c("24fM", "35aM","7bM","27eM","53bM","5aM")
readcnt_leaf_m <- readcnt_leaf_m %>%
  dplyr::select(-all_of(cols_to_remove)) # 71

cols_to_remove <- c("14bF", "40aF", "45aF", "67eF")
readcnt_leaf_f <- readcnt_leaf_f %>%
  dplyr::select(-all_of(cols_to_remove)) # 74

readcnt_leaf <- cbind(readcnt_leaf$Geneid, readcnt_leaf_m, readcnt_leaf_f)
readcnt_leaf_m <- cbind(readcnt_leaf[,1], readcnt_leaf_m)
readcnt_leaf_f <- cbind(readcnt_leaf[,1], readcnt_leaf_f)

# remove genes with low read count 
readcnt_leaf_m$all_mean <- rowMeans(readcnt_leaf_m[,c(2:72)])
readcnt_leaf_f$all_mean <- rowMeans(readcnt_leaf_f[,c(2:75)])

readcnt_leaf_m <- readcnt_leaf_m %>% filter(all_mean >= 5)  # 19348
readcnt_leaf_f <- readcnt_leaf_f %>% filter(all_mean >= 5)  # 18604
colnames(readcnt_leaf)[1] <- "gene"
colnames(readcnt_leaf_m)[1] <- "gene"
colnames(readcnt_leaf_f)[1] <- "gene"

write.table(readcnt_leaf, file = paste0(data,"readcnt_leaf_sept.txt"), row.names = FALSE, quote = F) # 37659
write.table(readcnt_leaf_m, file = paste0(data,"readcnt_leaf_m_sept.txt"), row.names = FALSE, quote = F)
write.table(readcnt_leaf_f, file = paste0(data,"readcnt_leaf_f_sept.txt"), row.names = FALSE, quote = F)
#readcnt_leaf_f <- read.table(paste0(data,"readcnt_leaf_f_sept.txt"), header = T, check.names = F)

# remove Y for FL
gff <- read.table(paste0(data,"merged_TX_noMatPARlarge_txanno_genes.gtf"))
gff <- gff[,c(13,1,4,5)]
colnames(gff) <- c("gene","chrom", "start", "end")

gff <- gff %>% mutate(chrom_new = case_when(
    chrom == "A1" ~ "A1",
    chrom == "A2" ~ "A2",
    chrom == "A3" ~ "A3",
    chrom == "A4" ~ "A4",
    chrom == "X" ~ "X",
    chrom == "Y" & start < 45000000 ~ "PAR",
    chrom == "Y" & start >= 45000000 ~ "Y"
))

readcnt_leaf_f <- inner_join(gff, readcnt_leaf_f, by = "gene")
# keep the X chromosome and PAR
readcnt_leaf_f <- readcnt_leaf_f %>% filter(chrom_new != "Y") # 14634 A 17630
table(readcnt_leaf_f$chrom_new)
write.table(readcnt_leaf_f, file = paste0(data,"readcnt_leaf_f_filt.txt"), row.names = FALSE, quote = F)
```
A1   A2   A3   A4  PAR    X 
4928 4668 2691 2347 1102 1894

## run deseq2 just to get normalized read counts
### FL
```{r}
# prep for deseq2
# exclude gene info columns and all_mean column
readcnt_matrix<-as.matrix(readcnt_leaf_f[ , -c(1:5,80)]) 
rownames(readcnt_matrix) <- readcnt_leaf_f[ , 1]
str(readcnt_matrix)
meta_data <- data.frame(colnames(readcnt_matrix))

colnames(meta_data) <- "id"
rownames(meta_data) <- meta_data[,1]
meta_data$sex <- "f"
head(meta_data)

# run deseq2
dds <- DESeqDataSetFromMatrix(countData = readcnt_matrix, 
                              colData = meta_data,
                              design = ~ 1)
dds <- DESeq(dds)
resultsNames(dds) # "Intercept"
res <- results(dds)
head(res)
summary(res)
result <- res[complete.cases(res),]
nrow(res) #  14804
nrow(result) # 14804

normalized_countsl <- as.data.frame(counts(dds, normalized=T))
normalized_countsl <- normalized_countsl %>% mutate(gene=rownames(normalized_countsl)) 
normalized_countsl <- normalized_countsl[,c(75,1:74)]

write.table(normalized_countsl, file = paste0(data,"normalized_counts_fl_sept.txt"), row.names = FALSE, quote = F)
normalized_counts_fl <- normalized_countsl
```

### ML
```{r}
# prep for deseq2
# exclude gene info columns and all_mean column
readcnt_matrix<-as.matrix(readcnt_leaf_m[ , -c(1,73)]) 
rownames(readcnt_matrix) <- readcnt_leaf_m[ , 1]
str(readcnt_matrix)
meta_data <- data.frame(colnames(readcnt_matrix))

colnames(meta_data) <- "id"
rownames(meta_data) <- meta_data[,1]
meta_data$sex <- "m"
head(meta_data)

# run deseq2
dds <- DESeqDataSetFromMatrix(countData = readcnt_matrix, 
                              colData = meta_data,
                              design = ~ 1)
dds <- DESeq(dds)
resultsNames(dds) # "Intercept"
res <- results(dds)
head(res)
summary(res)
result <- res[complete.cases(res),]
nrow(res) #  19348
nrow(result) # 19348

normalized_counts_ml <- as.data.frame(counts(dds, normalized=T))
normalized_counts_ml <- normalized_counts_ml %>% mutate(gene=rownames(normalized_counts_ml)) 
normalized_counts_ml <- normalized_counts_ml[,c(72,1:71)]
# includes sex chromosomes, separate autosomes later
write.table(normalized_counts_ml, file = paste0(data,"normalized_counts_ml_sept.txt"), row.names = FALSE, quote = F)
```


## quantile normalizatin
```{r}
qn <- function(x){qqnorm(x, plot.it=FALSE)$x}

# quantile normalize, now columns genes, rows are inds
normalized_counts_fln = apply(normalized_counts_fl[,-1], 1, qn)
normalized_counts_fln <- as.data.frame(t(normalized_counts_fln))
normalized_counts_fln <- cbind(normalized_counts_fl[,1], normalized_counts_fln)
colnames(normalized_counts_fln) <- colnames(normalized_counts_fl)

normalized_counts_mln = apply(normalized_counts_ml[,-1], 1, qn)
normalized_counts_mln <- as.data.frame(t(normalized_counts_mln))
normalized_counts_mln <- cbind(normalized_counts_ml[,1], normalized_counts_mln)
colnames(normalized_counts_mln) <- colnames(normalized_counts_ml)

```

## add TSS info
```{r}
gff_full <- read.table(paste0(data, "merged_TX_noMatPARlarge_txanno_gene_full.bed"))
colnames(gff_full) <- c("chrom", "start", "end", "strand")
gff_full <- inner_join(gff_full, gff, by = c("chrom", "start", "end"))
gff_full <- gff_full %>%
    mutate(start_tss = ifelse(strand == "+", start, end)) %>%
    mutate(end_tss = start_tss+1) 
gff_full <- gff_full[,c(6,7,8,5)]
colnames(gff_full) <- c("chrom", "start", "end", "gene")


normalized_counts_fln <- inner_join(gff_full, normalized_counts_fln, by = "gene")
#sort start pos for each chrom
normalized_counts_fln <- normalized_counts_fln %>%
  arrange(chrom, start) 

colnames(normalized_counts_fln)[4] <- "gene_id"
normalized_counts_fln$chrom <- factor(normalized_counts_fln$chrom)
levels(normalized_counts_fln$chrom)
normalized_counts_fln_X <- normalized_counts_fln %>% filter(chrom == "X") #1894
normalized_counts_fln_PAR <- normalized_counts_fln %>% filter(chrom == "PAR") #1102
normalized_counts_fln_auto <- normalized_counts_fln %>% filter(chrom != "PAR" & chrom != "X") #14634

colnames(normalized_counts_fln)[1] <- "#chr"
colnames(normalized_counts_fln_X)[1] <- "#chr"
colnames(normalized_counts_fln_PAR)[1] <- "#chr"
colnames(normalized_counts_fln_auto)[1] <- "#chr"

write.table(normalized_counts_fln, file = paste0(data,"normalized_counts_fln_X_PAR.bed"), row.names = FALSE, quote = F, sep = "\t")
write.table(normalized_counts_fln_X, file = paste0(data,"normalized_counts_fln_X.bed"), row.names = FALSE, quote = F, sep = "\t")
write.table(normalized_counts_fln_PAR, file = paste0(data,"normalized_counts_fln_PAR.bed"), row.names = FALSE, quote = F, sep = "\t")
write.table(normalized_counts_fln_auto, file = paste0(data,"normalized_counts_fln_auto.bed"), row.names = FALSE, quote = F, sep = "\t")

```

```{r}
# male leaf
normalized_counts_mln <- inner_join(gff_full, normalized_counts_mln, by = "gene")

#sort start pos for each chrom
normalized_counts_mln <- normalized_counts_mln %>%
  arrange(chrom, start) 

colnames(normalized_counts_mln)[4] <- "gene_id"
normalized_counts_mln$chrom <- factor(normalized_counts_mln$chrom)
table(normalized_counts_mln$chrom)

normalized_counts_mln_X <- normalized_counts_mln %>% filter(chrom == "X") # 1827
normalized_counts_mln_PAR <- normalized_counts_mln %>% filter(chrom == "PAR") # 1135
normalized_counts_mln_Y <- normalized_counts_mln %>% filter(chrom == "Y") # 1446
normalized_counts_mln_auto <- normalized_counts_mln %>% filter(chrom != "PAR" & chrom != "X" & chrom != "Y") # 14940

colnames(normalized_counts_mln)[1] <- "#chr"
colnames(normalized_counts_mln_X)[1] <- "#chr"
colnames(normalized_counts_mln_PAR)[1] <- "#chr"
colnames(normalized_counts_mln_Y)[1] <- "#chr"
colnames(normalized_counts_mln_auto)[1] <- "#chr"

write.table(normalized_counts_mln, file = paste0(data,"normalized_counts_mln_full.bed"), row.names = FALSE, quote = F, sep = "\t")
write.table(normalized_counts_mln_X, file = paste0(data,"normalized_counts_mln_X.bed"), row.names = FALSE, quote = F, sep = "\t")
write.table(normalized_counts_mln_PAR, file = paste0(data,"normalized_counts_mln_PAR.bed"), row.names = FALSE, quote = F, sep = "\t")
write.table(normalized_counts_mln_Y, file = paste0(data,"normalized_counts_mln_Y.bed"), row.names = FALSE, quote = F, sep = "\t")
write.table(normalized_counts_mln_auto, file = paste0(data,"normalized_counts_mln_auto.bed"), row.names = FALSE, quote = F, sep = "\t")
```
A1   A2   A3   A4  PAR    X    Y 
5041 4770 2739 2390 1135 1827 1446
