---
title: "pollen leaf read count difference"
author: "Meng"
date: "2025-02-04"
output: html_document
---

```{r setup, include=FALSE}
library(arrow)
library(dplyr)
library(tidyr)
library(ggplot2)
library(qvalue)
data <- "/Users/yuanmeng/Library/CloudStorage/OneDrive-UniversityofToronto/Manuscripts/rumex_eqtl/eqtl/"
```

## get eQTL from ML MP for comparison
replace ML with ML_flipped
```{r}
ML_flipped <- read.table(paste0(data, "ML_flipped.cis_qtl.txt"), header = T) # 14777
MP_flipped <- read.table(paste0(data, "MP_flipped.cis_qtl.txt"), header = T) # 13785
ML_flipped_egene <- ML_flipped %>% filter(qval < 0.1) # 2427
MP_flipped_egene <- MP_flipped %>% filter(qval < 0.1) # 1480

ML_flipped_A1 <- read_parquet(paste0(data,"ML_flipped.cis_qtl_pairs.1.parquet"))
ML_flipped_A2 <- read_parquet(paste0(data,"ML_flipped.cis_qtl_pairs.2.parquet"))
ML_flipped_A3 <- read_parquet(paste0(data,"ML_flipped.cis_qtl_pairs.3.parquet"))
ML_flipped_A4 <- read_parquet(paste0(data,"ML_flipped.cis_qtl_pairs.4.parquet"))
ML_flipped_eqtl <- rbind(ML_flipped_A1, ML_flipped_A2, ML_flipped_A3, ML_flipped_A4) # 1272993
sum(ML_flipped$num_var) # 1272993

MP_flipped_A1 <- read_parquet(paste0(data,"MP_flipped.cis_qtl_pairs.1.parquet"))
MP_flipped_A2 <- read_parquet(paste0(data,"MP_flipped.cis_qtl_pairs.2.parquet"))
MP_flipped_A3 <- read_parquet(paste0(data,"MP_flipped.cis_qtl_pairs.3.parquet"))
MP_flipped_A4 <- read_parquet(paste0(data,"MP_flipped.cis_qtl_pairs.4.parquet"))
MP_flipped_eqtl <- rbind(MP_flipped_A1, MP_flipped_A2, MP_flipped_A3, MP_flipped_A4) # 1182102
sum(MP_flipped$num_var) # 1182102

ML_flipped_eqtl_full <- inner_join(ML_flipped_eqtl, ML_flipped_egene[,c(1,16:19)], by = "phenotype_id") # 262510
ML_eqtl_sig <- ML_flipped_eqtl_full %>% filter(pval_nominal < pval_nominal_threshold) # 16617
MP_flipped_eqtl_full <- inner_join(MP_flipped_eqtl, MP_flipped_egene[,c(1,16:19)], by = "phenotype_id") # 160522
MP_eqtl_sig <- MP_flipped_eqtl_full %>% filter(pval_nominal < pval_nominal_threshold) # 9463
```


# get eQTL for expression difference between pollen and leaf
## permutations
```{r}
# pl <- read.table(paste0(data, "pl.cis_qtl.txt"), header = T) # 16218
pl <- read.table(paste0(data, "pl_abs.cis_qtl.txt"), header = T) # 16218

qval_pl <- qvalue(pl$pval_beta)
summary(qval_pl)
hist(qval_pl)
```

## extract eQenes and eQTLs
```{r}
pl_egene <- pl %>% filter(qval < 0.1) # 1566

pl_A1 <- read_parquet(paste0(data,"pl_abs.cis_qtl_pairs.1.parquet"))
pl_A2 <- read_parquet(paste0(data,"pl_abs.cis_qtl_pairs.2.parquet"))
pl_A3 <- read_parquet(paste0(data,"pl_abs.cis_qtl_pairs.3.parquet"))
pl_A4 <- read_parquet(paste0(data,"pl_abs.cis_qtl_pairs.4.parquet"))
pl_eqtl <- rbind(pl_A1, pl_A2, pl_A3, pl_A4) # 1404801
sum(pl$num_var) # 1404801
```

```{r}
ggplot(pl_eqtl, aes(x=pval_nominal)) + geom_histogram(color="black", fill="white",binwidth = 0.01) + theme_classic()+ xlab("p-value")+ggtitle("")
```

```{r}
pl_eqtl_full <- inner_join(pl_eqtl, pl_egene[,c(1,16:19)], by = "phenotype_id") # 176402
pl_eqtl_sig <- pl_eqtl_full %>% filter(pval_nominal < pval_nominal_threshold) # 9873

length(unique(pl_eqtl_sig$variant_id)) # 9639
length(unique(pl_eqtl_sig$phenotype_id)) # 1564

pl_eqtl_sig <- pl_eqtl_sig %>% mutate(maf = ifelse(af <= 0.5, af, 1 - af))
write.csv(pl_eqtl_sig, file = paste0(data, "pl_eqtl_sig.csv"), row.names = F, quote = F)
pl_eqtl_sig <- read.csv(paste0(data, "pl_eqtl_sig.csv"))

set.seed(123)
pl_eqtl_sig_random <- pl_eqtl_sig %>%
  group_by(phenotype_id) %>% 
  slice_sample(n = 1)  # 1564
write.csv(pl_eqtl_sig_random, file = paste0(data, "pl_eqtl_sig_random.csv"), row.names = F, quote = F)

pl_eqtl_sig_min <- pl_eqtl_sig %>%
  group_by(phenotype_id) %>% 
  filter(pval_nominal == min(pval_nominal)) %>% 
  slice_sample(n = 1) # 1564
write.csv(pl_eqtl_sig_min, file = paste0(data, "pl_eqtl_sig_min.csv"), row.names = F, quote = F)
```


## MAF
```{r}
get_hist_prop <- function(data) {
  p <- ggplot(data, aes(x = maf)) + geom_histogram(binwidth = 0.01,stat="count")
  p_data <- ggplot_build(p)$data[[1]]
  p_data$y_prop <- p_data$y / sum(p_data$y)
  return(p_data)
}

pl_eqtl_sig_min_af <- pl_eqtl_sig_min %>% dplyr::select(variant_id, af, maf) %>% distinct()
pl_eqtl_sig_random_af <- pl_eqtl_sig_random %>% dplyr::select(variant_id, af, maf) %>% distinct()

p<-ggplot(pl_eqtl_sig_min_af, aes(x=maf)) + geom_histogram(binwidth=0.01) 
p_data <- ggplot_build(p)$data[[1]]
p_data$y_prop <- p_data$y/sum(p_data$y) # 0.111
p1 <- ggplot(p_data, aes(x=x, y=y_prop)) +
  geom_bar(stat="identity", width = 0.01, color="white")  + theme_classic() + ggtitle("a")+ labs(x="MAF", y="Proportion")+ theme(text = element_text(size = 10)) + ylim(0, 0.13)

p<-ggplot(pl_eqtl_sig_random_af, aes(x=maf)) + geom_histogram(binwidth=0.01) 
p_data <- ggplot_build(p)$data[[1]]
p_data$y_prop <- p_data$y/sum(p_data$y) # 0.111
p2 <- ggplot(p_data, aes(x=x, y=y_prop)) +
  geom_bar(stat="identity", width = 0.01, color="white")  + theme_classic() + ggtitle("b")+ labs(x="MAF", y="Proportion")+ theme(text = element_text(size = 10)) + ylim(0, 0.13)
p1 + p2
```


## distribution of effect size

```{r}
nrow(pl_eqtl_sig[pl_eqtl_sig$slope > 0,]) # 6173
nrow(pl_eqtl_sig[pl_eqtl_sig$slope < 0,]) # 3700

nrow(pl_eqtl_sig_min[pl_eqtl_sig_min$slope > 0,]) # 921
nrow(pl_eqtl_sig_min[pl_eqtl_sig_min$slope < 0,]) # 643

nrow(pl_eqtl_sig_random[pl_eqtl_sig_random$slope > 0,]) # 931
nrow(pl_eqtl_sig_random[pl_eqtl_sig_random$slope < 0,]) # 633
```

```{r}
p3 <- ggplot(pl_eqtl_sig, aes(x=slope)) + geom_histogram(color="black", fill="white",binwidth = 0.01) + theme_classic()+ xlab("effect size")+ggtitle("")

p4 <- ggplot(pl_eqtl_sig_min, aes(x=slope)) + geom_histogram(color="black", fill="white",binwidth = 0.01) + theme_classic()+ xlab("effect size")+ggtitle("")
p3+p4
ggsave(paste0(data,"pl_eqtl_beta.pdf"), width = 7, height = 3, units = "in")
```



## cis independent
1566 eGenes
```{r}
pl_independent <- read.table(paste0(data, "pl_abs.cis_independent_qtl.txt"), header = T) # 1707
egene <- as.data.frame(unique(pl_eqtl_sig$phenotype_id))
colnames(egene) <- "phenotype_id"
pl_independent <- left_join(egene, pl_independent, by = "phenotype_id") # 1705
length(unique(pl_independent$phenotype_id)) # 1564

pl1 <- pl_independent %>% filter(rank == 1) # 1559

# filter for primary eQTLs
pl1 <- pl_independent %>%
  group_by(phenotype_id) %>% 
  filter(rank == min(rank)) # 1564

# manually update the top rank for each eGene to rank = 1
pl1$rank <- 1 
# filter for secondary eQTLs
pl2 <- anti_join(pl_independent, pl1, by = c("phenotype_id", "variant_id")) # 141
length(pl2$phenotype_id)

pl1 <- pl1 %>% dplyr::select(phenotype_id, variant_id, pval_nominal, slope, rank)
pl2 <- pl2 %>% dplyr::select(phenotype_id, variant_id, pval_nominal, slope, rank)
test <- inner_join(pl1, pl2, by = "phenotype_id")
test1 <- test %>% filter(slope.x * slope.y < 0) # 54
```

## eqtls affecting multiple genes
```{r}
length(unique(pl_eqtl_sig$variant_id)) # 9639
pl_eqtl_multigene <- pl_eqtl_sig %>% 
     group_by(variant_id) %>%
     filter(n()>1) %>%
    dplyr::select(variant_id, maf) %>% unique() # 231

pl_eqtl_onegene <- pl_eqtl_sig %>% 
     group_by(variant_id) %>%
     filter(n()==1) %>%
    dplyr::select(variant_id, maf) # 9408

pl_eqtl_multigene <- left_join(pl_eqtl_multigene, pl_eqtl_sig, by = c("variant_id", "maf")) # 465
length(unique(pl_eqtl_multigene$phenotype_id)) # 101
```


# check overlap with results from separate models
## overlap with eQTLs in ML, MP
```{r}
ml_overlap <- inner_join(pl_eqtl_sig, ML_eqtl_sig, by = c("phenotype_id", "variant_id")) # 4538
mp_overlap <- inner_join(pl_eqtl_sig, MP_eqtl_sig, by = c("phenotype_id", "variant_id")) # 4977

no_overlap <- pl_eqtl_sig %>% anti_join(ML_eqtl_sig, by = c("phenotype_id", "variant_id")) %>% anti_join(MP_eqtl_sig, by = c("phenotype_id", "variant_id")) # 1125
```

```{r}
data1 <- "/Users/yuanmeng/Library/CloudStorage/OneDrive-UniversityofToronto/Manuscripts/rumex_eqtl/"
normalized_counts_mln <- read.table(paste0(data1,"normalized_counts_mln.bed"), header = T, check.names = F)
normalized_counts_mpn <- read.table(paste0(data1,"normalized_counts_mpn.bed"), header = T, check.names = F)

# genes tested in all MP, ML and pl_diff 1051
ml_gene <- as.data.frame(normalized_counts_mln[,4])
mp_gene <- as.data.frame(normalized_counts_mpn[,4])
colnames(ml_gene) <- "phenotype_id"
colnames(mp_gene) <- "phenotype_id"
overlap_gene <- pl_eqtl_sig %>% inner_join(ml_gene, by = "phenotype_id") %>% inner_join(mp_gene, by = "phenotype_id") %>% dplyr::select(phenotype_id, variant_id) %>% distinct() # 5946 eqtls
length(unique(overlap_gene$phenotype_id)) # 1051

# for Venn diagram
write.table(overlap_gene, file = paste0(data, "pl_gene_overlap.txt"), row.names = F, quote = F, col.names = F)

no_overlap2 <- overlap_gene %>% anti_join(ML_eqtl_sig, by = c("phenotype_id", "variant_id")) %>% anti_join(MP_eqtl_sig, by = c("phenotype_id", "variant_id")) # 919
5946 - 919

# check signs of beta of shared eqtls
pl <- pl_eqtl_sig %>% dplyr::select(phenotype_id, variant_id, slope)
ml <- ML_eqtl_sig %>% dplyr::select(phenotype_id, variant_id, slope)
mp <- MP_eqtl_sig %>% dplyr::select(phenotype_id, variant_id, slope)
colnames(pl)[3] <- "pl"
colnames(ml)[3] <- "ml"
colnames(mp)[3] <- "mp"

overlap_eqtl <- pl %>% inner_join(ml, by = c("phenotype_id", "variant_id")) %>% inner_join(mp, by = c("phenotype_id", "variant_id")) # 767

# different degree
sum(overlap_eqtl$pl > 0 & overlap_eqtl$ml > 0 & overlap_eqtl$mp > 0) # 6
sum(overlap_eqtl$pl > 0 & overlap_eqtl$ml < 0 & overlap_eqtl$mp < 0) # 522
sum(overlap_eqtl$pl > 0 & overlap_eqtl$ml > 0 & overlap_eqtl$mp < 0) # 22
sum(overlap_eqtl$pl > 0 & overlap_eqtl$ml < 0 & overlap_eqtl$mp > 0) # 0

sum(overlap_eqtl$pl < 0 & overlap_eqtl$ml > 0 & overlap_eqtl$mp > 0) # 109
sum(overlap_eqtl$pl < 0 & overlap_eqtl$ml < 0 & overlap_eqtl$mp < 0) # 98
sum(overlap_eqtl$pl < 0 & overlap_eqtl$ml > 0 & overlap_eqtl$mp < 0) # 1
sum(overlap_eqtl$pl < 0 & overlap_eqtl$ml < 0 & overlap_eqtl$mp > 0) # 9

# pl and leaf same sign
6+98+22+9

# pl and pollen same sign
6+98+0+1

# leaf and pollen same sign
6+522+109+98 # 735s
```


## overlap with eQTLs show 'G x Lifestage'
redo using the flipped results
```{r}
dis_pl <- read.csv(paste0(data, "dis_pl.csv")) # 974
overlap <- inner_join(pl_eqtl_sig, dis_pl, by = c("phenotype_id", "variant_id")) # 479
table(overlap$category)

dis_pl <- dis_pl %>% filter(category == "both" | category == "leaf" | category == "pollen") # 19
overlap <- inner_join(pl_eqtl_sig, dis_pl, by = c("phenotype_id", "variant_id")) # 19

dis_pl <- dis_pl %>% filter(category == "leaf2" | category == "pollen2") # 955
overlap <- inner_join(pl_eqtl_sig, dis_pl, by = c("phenotype_id", "variant_id")) # 460

```
both    leaf   leaf2  pollen pollen2 
      9       5     184       5     276 2



# overlap between eGenes and DE genes from deseq2
```{r}
DE_tissue <- read.table("/Users/yuanmeng/Library/CloudStorage/OneDrive-UniversityofToronto/Manuscripts/rumex_eqtl/DE_tissue_full.txt", header = T)
head(DE_tissue) 
DE_tissue <- DE_tissue %>% filter(padj<0.05) # 19822
egene <- as.data.frame(unique(pl_eqtl_sig$phenotype_id)) # 1564
colnames(egene) <- "gene" 
egene_overlap <- inner_join(egene, DE_tissue, by = "gene") # 1464

# DE_tissue <- DE_tissue %>% filter(padj<0.1) # 200061
# egene_overlap <- inner_join(egene, DE_tissue, by = "gene") # 1905

p1 <- ggplot(DE_tissue, aes(abs(log2FoldChange))) + geom_histogram(binwidth = 0.5,color="black", fill="white") + theme_classic() + ggtitle("a")

p2 <- ggplot(egene_overlap, aes(abs(log2FoldChange))) + geom_histogram(binwidth = 0.5,color="black", fill="white") + theme_classic() + ggtitle("b")

p1 + p2
ggsave(paste0(data,"logFC_distribution_both.pdf"), width = 7, height = 3, units = "in")

```


```{r}
DE_tissue <- DE_tissue %>% mutate(bias = case_when(
    abs(log2FoldChange) <= 1 ~ "FC1",
    abs(log2FoldChange) > 1 & abs(log2FoldChange) <= 2 ~ "FC2",
    abs(log2FoldChange) > 2 & abs(log2FoldChange) <= 3 ~ "FC3",
    abs(log2FoldChange) > 3 & abs(log2FoldChange)  ~ "FC4"
))
t1 <- as.data.frame(table(DE_tissue$bias))

egene_overlap <- egene_overlap %>% mutate(bias = case_when(
    abs(log2FoldChange) <= 1 ~ "FC1",
    abs(log2FoldChange) > 1 & abs(log2FoldChange) <= 2 ~ "FC2",
    abs(log2FoldChange) > 2 & abs(log2FoldChange) <= 3 ~ "FC3",
    abs(log2FoldChange) > 3 & abs(log2FoldChange)  ~ "FC4"
))
t2 <- as.data.frame(table(egene_overlap$bias))

colnames(t1) <- c("bias","all")
colnames(t2) <- c("bias","egene")
t <- inner_join(t1, t2, by = "bias")
rownames(t) <- t[,1]
t <- t[,-1]
chi <- chisq.test(t)
chi$residuals
mosaicplot(t,shade = TRUE)
ggsave(paste0(data,"mosaicplot_residuals.pdf"), width = 7, height = 3, units = "in")

```
Pearson's Chi-squared test

data:  t
X-squared = 44.871, df = 3, p-value = 9.858e-10
           all     egene
FC1  1.3442574 -4.946359
FC2  0.4706882 -1.731955
FC3 -0.4691880  1.726434
FC4 -0.9150820  3.367156

```{r}
median(abs(egene_overlap$log2FoldChange)) # 2.607984
summary(abs(egene_overlap$log2FoldChange))
```
 Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
 0.1234  1.6093  2.6080  3.4807  4.0159 28.2721 
 
# plot eQTLs for DE between ML and MP
## TX_paternal_00029854 4:34118808
```{r}
expr_l <- normalized_counts_mln %>% filter(gene_id == "TX_paternal_00029854")
expr_p <- normalized_counts_mpn %>% filter(gene_id == "TX_paternal_00029854")

geno <- read.table(paste0(data,"A4_34118808.vcf"), header = T, check.names = F)
geno <- as.data.frame(t(geno[,c(5:73)]))
geno <- geno %>% mutate(id = row.names(geno))
colnames(geno) <- c("genotype", "id")

expr_l <- as.data.frame(t(expr_l[,c(5:79)]))
expr_p <- as.data.frame(t(expr_p[,c(5:77)]))
expr_l <- expr_l %>% mutate(id = row.names(expr_l))
colnames(expr_l) <- c("leaf", "id")
expr_p <- expr_p %>% mutate(id = row.names(expr_p))
colnames(expr_p) <- c("pollen", "id")

eqtl_lp1 <- geno %>% left_join(expr_l, by = "id") %>% left_join(expr_p, by = "id")
eqtl_lp1$genotype <- as.factor(eqtl_lp1$genotype)
table(eqtl_lp1$genotype)
write.table(eqtl_lp1, file = paste0(data, "A4_34118808_TX_paternal_00029854.txt"), quote = F)
eqtl_lp1 <- gather(eqtl_lp1, lifestage, Expression, leaf:pollen)

levels(eqtl_lp1$genotype) <- c("./.","TT", "TC")
eqtl_lp1 <- eqtl_lp1 %>% filter(genotype != "./.")
write.table(eqtl_lp1, file = paste0(data, "A4_34118808_TX_paternal_00029854_final.txt"), quote = F)
#eqtl_lp1 <- read.table(paste0(data, "eqtl_lp1_final.txt"))
ggplot(eqtl_lp1, aes(x=genotype, y=Expression, colour=lifestage)) + 
  geom_boxplot(position=position_dodge2(0.85)) + 
  geom_jitter(shape=16, position=position_jitterdodge(0.2), alpha=0.3) + 
  theme_classic() + theme(text = element_text(size = 12))+ scale_colour_manual(values = c("red", "blue"))+labs(x="Genotype", y="Expression")
```
./. 0/0 0/1 
 12  47  10
 

