---
title: "pollen leaf read count difference"
author: "Meng"
date: "2025-02-04"
output: html_document
---

```{r setup, include=FALSE}
library(arrow)
library(dplyr)
library(tidyr)
library(ggplot2)
library(qvalue)
data <- "/Users/yuanmeng/Library/CloudStorage/OneDrive-UniversityofToronto/Manuscripts/rumex_eqtl/eqtl/"
```

```{r}
ML_eqtl_sig <-read.csv(paste0(data, "ML_eqtl_sig_Nov.csv"))
MP_eqtl_sig <-read.csv(paste0(data, "MP_eqtl_sig_Nov.csv"))
```


# get eQTL for expression difference between pollen and leaf
## permutations
```{r}
pl <- read.table(paste0(data, "pl.cis_qtl.txt"), header = T) # 16218

qval_pl <- qvalue(pl$pval_beta)
summary(qval_pl)
hist(qval_pl)
```

## extract eQenes and eQTLs
```{r}
pl_egene <- pl %>% filter(qval < 0.1) # 2031

pl_A1 <- read_parquet(paste0(data,"pl.cis_qtl_pairs.1.parquet"))
pl_A2 <- read_parquet(paste0(data,"pl.cis_qtl_pairs.2.parquet"))
pl_A3 <- read_parquet(paste0(data,"pl.cis_qtl_pairs.3.parquet"))
pl_A4 <- read_parquet(paste0(data,"pl.cis_qtl_pairs.4.parquet"))
pl_eqtl <- rbind(pl_A1, pl_A2, pl_A3, pl_A4) # 1404801
sum(pl$num_var) # 1404801
```

```{r}
ggplot(pl_eqtl, aes(x=pval_nominal)) + geom_histogram(color="black", fill="white",binwidth = 0.01) + theme_classic()+ xlab("p-value")+ggtitle("")
```

```{r}
pl_eqtl_full <- inner_join(pl_eqtl, pl_egene[,c(1,16:19)], by = "phenotype_id") # 223329
pl_eqtl_sig <- pl_eqtl_full %>% filter(pval_nominal < pval_nominal_threshold) # 12978

length(unique(pl_eqtl_sig$variant_id)) # 12626
length(unique(pl_eqtl_sig$phenotype_id)) # 2029

pl_eqtl_sig <- pl_eqtl_sig %>% mutate(maf = ifelse(af <= 0.5, af, 1 - af))
write.csv(pl_eqtl_sig, file = paste0(data, "pl_eqtl_sig.csv"), row.names = F, quote = F)
pl_eqtl_sig <- read.csv(paste0(data, "pl_eqtl_sig.csv"))

set.seed(123)
pl_eqtl_sig_random <- pl_eqtl_sig %>%
  group_by(phenotype_id) %>% 
  slice_sample(n = 1)  # 2029
write.csv(pl_eqtl_sig_random, file = paste0(data, "pl_eqtl_sig_random.csv"), row.names = F, quote = F)

pl_eqtl_sig_min <- pl_eqtl_sig %>%
  group_by(phenotype_id) %>% 
  filter(pval_nominal == min(pval_nominal)) %>% 
  slice_sample(n = 1) # 2029
write.csv(pl_eqtl_sig_min, file = paste0(data, "pl_eqtl_sig_min.csv"), row.names = F, quote = F)
```

## MAF
```{r}
get_hist_prop <- function(data) {
  p <- ggplot(data, aes(x = maf)) + geom_histogram(binwidth = 0.01,stat="count")
  p_data <- ggplot_build(p)$data[[1]]
  p_data$y_prop <- p_data$y / sum(p_data$y)
  return(p_data)
}

pl_eqtl_sig_min_af <- pl_eqtl_sig_min %>% dplyr::select(variant_id, af, maf) %>% distinct()
pl_eqtl_sig_random_af <- pl_eqtl_sig_random %>% dplyr::select(variant_id, af, maf) %>% distinct()

p<-ggplot(pl_eqtl_sig_min_af, aes(x=maf)) + geom_histogram(binwidth=0.01) 
p_data <- ggplot_build(p)$data[[1]]
p_data$y_prop <- p_data$y/sum(p_data$y) # 0.111
p1 <- ggplot(p_data, aes(x=x, y=y_prop)) +
  geom_bar(stat="identity", width = 0.01, color="white")  + theme_classic() + ggtitle("a")+ labs(x="MAF", y="Proportion")+ theme(text = element_text(size = 10)) + ylim(0, 0.13)

p<-ggplot(pl_eqtl_sig_random_af, aes(x=maf)) + geom_histogram(binwidth=0.01) 
p_data <- ggplot_build(p)$data[[1]]
p_data$y_prop <- p_data$y/sum(p_data$y) # 0.111
p2 <- ggplot(p_data, aes(x=x, y=y_prop)) +
  geom_bar(stat="identity", width = 0.01, color="white")  + theme_classic() + ggtitle("b")+ labs(x="MAF", y="Proportion")+ theme(text = element_text(size = 10)) + ylim(0, 0.13)
p1 + p2
```


## distribution of effect size

```{r}
nrow(pl_eqtl_sig[pl_eqtl_sig$slope > 0,]) # 5481
nrow(pl_eqtl_sig[pl_eqtl_sig$slope < 0,]) # 7497
```

```{r}
ggplot(p_data, aes(x=x, y=y_prop)) +
  geom_bar(stat="identity", width = 0.01, color="white")  + theme_classic() + ggtitle("b")+ labs(x="MAF", y="Proportion")+ theme(text = element_text(size = 10)) + ylim(0, 0.13)

p3 <- ggplot(pl_eqtl_sig, aes(x=slope)) + geom_histogram(color="black", fill="white",binwidth = 0.01) + theme_classic()+ xlab("effect size")+ggtitle("")

p4 <- ggplot(pl_eqtl_sig_min, aes(x=slope)) + geom_histogram(color="black", fill="white",binwidth = 0.01) + theme_classic()+ xlab("effect size")+ggtitle("")
p3+p4
```




## cis independent
2029 eGenes
```{r}
pl_independent <- read.table(paste0(data, "pl.cis_independent_qtl.txt"), header = T) # 2218
length(unique(pl_independent$phenotype_id)) # 2029

pl1 <- pl_independent %>% filter(rank == 1) # 2020

# filter for primary eQTLs
pl1 <- pl_independent %>%
  group_by(phenotype_id) %>% 
  filter(rank == min(rank)) # 2029

# manually update the top rank for each eGene to rank = 1
pl1$rank <- 1 # 2029
# filter for secondary eQTLs
pl2 <- anti_join(pl_independent, pl1, by = c("phenotype_id", "variant_id")) # 188
length(pl2$phenotype_id)

pl1 <- pl1 %>% select(phenotype_id, variant_id, pval_nominal, slope, rank)
pl2 <- pl2 %>% select(phenotype_id, variant_id, pval_nominal, slope, rank)
test <- inner_join(pl1, pl2, by = "phenotype_id")
test1 <- test %>% filter(slope.x * slope.y < 0) #72
# rank = 1 but not the most significant eQTL
# test <- anti_join(pl1, pl_egene, by = c("phenotype_id", "variant_id")) # 30

```

## eqtls affecting multiple genes
```{r}
length(unique(pl_eqtl_sig$variant_id)) # 12626
pl_eqtl_multigene <- pl_eqtl_sig %>% 
     group_by(variant_id) %>%
     filter(n()>1) %>%
    select(variant_id, maf) %>% unique() # 343

pl_eqtl_onegene <- pl_eqtl_sig %>% 
     group_by(variant_id) %>%
     filter(n()==1) %>%
    select(variant_id, maf) # 12283
```


# check overlap with results from separate models
## overlap with eQTLs in ML, MP
```{r}
ml_overlap <- inner_join(pl_eqtl_sig, ML_eqtl_sig, by = c("phenotype_id", "variant_id")) # 4986
mp_overlap <- inner_join(pl_eqtl_sig, MP_eqtl_sig, by = c("phenotype_id", "variant_id")) # 6153

no_overlap <- pl_eqtl_sig %>% anti_join(ML_eqtl_sig, by = c("phenotype_id", "variant_id")) %>% anti_join(MP_eqtl_sig, by = c("phenotype_id", "variant_id")) # 2534
# TX_paternal_00028051 3:98850390

# for Venn diagram
write.table(pl_eqtl_sig[,c(1:2)], file = paste0(data, "pl_eqtl.txt"), row.names = F, quote = F, col.names = F)
```

```{r}
# for Venn diagram
write.table(normalized_counts_mln[,4], file = paste0(data, "ml_gene.txt"), row.names = F, quote = F, col.names = F)
write.table(normalized_counts_mpn[,4], file = paste0(data, "mp_gene.txt"), row.names = F, quote = F, col.names = F)
write.table(pl_eqtl_sig[,1], file = paste0(data, "pl_gene.txt"), row.names = F, quote = F, col.names = F)

# genes tested in all MP, ML and pl_diff 1450
ml_gene <- as.data.frame(normalized_counts_mln[,4])
mp_gene <- as.data.frame(normalized_counts_mpn[,4])
colnames(ml_gene) <- "phenotype_id"
colnames(mp_gene) <- "phenotype_id"
overlap_gene <- pl_eqtl_sig %>% inner_join(ml_gene, by = "phenotype_id") %>% inner_join(mp_gene, by = "phenotype_id") %>% select(phenotype_id, variant_id) %>% distinct() # 1450 egenes, 8582 eqtls

# for Venn diagram
write.table(overlap_gene, file = paste0(data, "pl_gene_overlap.txt"), row.names = F, quote = F, col.names = F)

no_overlap2 <- overlap_gene %>% anti_join(ML_eqtl_sig, by = c("phenotype_id", "variant_id")) %>% anti_join(MP_eqtl_sig, by = c("phenotype_id", "variant_id")) # 1800
no_overlap2 <- no_overlap2 %>% inner_join(pl_eqtl_sig, by = c("phenotype_id", "variant_id"))
# TX_paternal_00029854 4:34118808
```
genes only tested in pl (from venn diagram)
TX_paternal_00004400 TX_paternal_00031744 TX_paternal_00006959 TX_paternal_00015600 TX_paternal_00028328 TX_paternal_00033344 TX_paternal_00006794 TX_paternal_00003125 TX_paternal_00002744 TX_paternal_00012841 TX_paternal_00030200


## overlap with eQTLs show 'G x Lifestage'
```{r}
dis_pl <- read.csv(paste0(data, "dis_pl.csv")) # 974
overlap <- inner_join(pl_eqtl_sig, dis_pl, by = c("phenotype_id", "variant_id")) # 479
table(overlap$category)

dis_pl <- dis_pl %>% filter(category == "both" | category == "leaf" | category == "pollen") # 19
overlap <- inner_join(pl_eqtl_sig, dis_pl, by = c("phenotype_id", "variant_id")) # 19

dis_pl <- dis_pl %>% filter(category == "leaf2" | category == "pollen2") # 955
overlap <- inner_join(pl_eqtl_sig, dis_pl, by = c("phenotype_id", "variant_id")) # 460

```
both    leaf   leaf2  pollen pollen2 
      9       5     184       5     276 2


# overlap between eGenes and DE genes from deseq2
```{r}
DE_tissue <- read.table("/Users/yuanmeng/Library/CloudStorage/OneDrive-UniversityofToronto/Manuscripts/rumex_eqtl/DE_tissue_full.txt", header = T)
head(DE_tissue) 
DE_tissue <- DE_tissue %>% filter(padj<0.05) # 19822
egene <- as.data.frame(unique(pl_eqtl_sig$phenotype_id)) # 2029
colnames(egene) <- "gene" 
egene_overlap <- inner_join(egene, DE_tissue, by = "gene") # 1874

DE_tissue <- DE_tissue %>% filter(padj<0.1) # 200061
egene_overlap <- inner_join(egene, DE_tissue, by = "gene") # 1905
```


# plot eQTLs for DE between ML and MP
## TX_paternal_00029854 4:34118808
```{r}
data <- "/Users/yuanmeng/Library/CloudStorage/OneDrive-UniversityofToronto/Manuscripts/rumex_eqtl/"
normalized_counts_mln <- read.table(paste0(data,"normalized_counts_mln.bed"), header = T, check.names = F)
normalized_counts_mpn <- read.table(paste0(data,"normalized_counts_mpn.bed"), header = T, check.names = F)

expr_l <- normalized_counts_mln %>% filter(gene_id == "TX_paternal_00029854")
expr_p <- normalized_counts_mpn %>% filter(gene_id == "TX_paternal_00029854")

geno <- read.table(paste0(data,"A4_34118808.vcf"), header = T, check.names = F)
geno <- as.data.frame(t(geno[,c(5:73)]))
geno <- geno %>% mutate(id = row.names(geno))
colnames(geno) <- c("genotype", "id")

expr_l <- as.data.frame(t(expr_l[,c(5:79)]))
expr_p <- as.data.frame(t(expr_p[,c(5:77)]))
expr_l <- expr_l %>% mutate(id = row.names(expr_l))
colnames(expr_l) <- c("leaf", "id")
expr_p <- expr_p %>% mutate(id = row.names(expr_p))
colnames(expr_p) <- c("pollen", "id")

eqtl_lp1 <- geno %>% left_join(expr_l, by = "id") %>% left_join(expr_p, by = "id")
eqtl_lp1$genotype <- as.factor(eqtl_lp1$genotype)
table(eqtl_lp1$genotype)
write.table(eqtl_lp1, file = paste0(data, "A4_34118808_TX_paternal_00029854.txt"), quote = F)
eqtl_lp1 <- gather(eqtl_lp1, lifestage, Expression, leaf:pollen)

levels(eqtl_lp1$genotype) <- c("./.","TT", "TC")
eqtl_lp1 <- eqtl_lp1 %>% filter(genotype != "./.")
write.table(eqtl_lp1, file = paste0(data, "A4_34118808_TX_paternal_00029854_final.txt"), quote = F)
#eqtl_lp1 <- read.table(paste0(data, "eqtl_lp1_final.txt"))
ggplot(eqtl_lp1, aes(x=genotype, y=Expression, colour=lifestage)) + 
  geom_boxplot(position=position_dodge2(0.85)) + 
  geom_jitter(shape=16, position=position_jitterdodge(0.2), alpha=0.3) + 
  theme_classic() + theme(text = element_text(size = 12))+ scale_colour_manual(values = c("red", "blue"))+labs(x="Genotype", y="Expression")
```
./. 0/0 0/1 
 12  47  10
 
direction of the slope
positive slope, negative slope, but not in discordance
```{r}

```


