---
title: "addtional info"
author: "Meng"
date: "2024-06-20"
output: html_document
---

```{r setup, include=FALSE}
library(dplyr)
library(ggplot2)
library(patchwork)
data <- "/Users/yuanmeng/Library/CloudStorage/OneDrive-UniversityofToronto/Manuscripts/rumex_eqtl/eqtl/"
ML_eqtl_sig <- read.csv(paste0(data, "ML_eqtl_sig_Nov.csv"))
FL_eqtl_sig <- read.csv(paste0(data, "FL_eqtl_sig_Nov.csv"))
MP_eqtl_sig <- read.csv(paste0(data, "MP_eqtl_sig_Nov.csv"))

ML_eqtl_sig_min <- read.csv(paste0(data, "ML_eqtl_sig_min.csv"))
FL_eqtl_sig_min <- read.csv(paste0(data, "FL_eqtl_sig_min.csv"))
MP_eqtl_sig_min <- read.csv(paste0(data, "MP_eqtl_sig_min.csv"))

ML_eqtl_sig_random <- read.csv(paste0(data, "ML_eqtl_sig_random.csv"))
FL_eqtl_sig_random <- read.csv(paste0(data, "FL_eqtl_sig_random.csv"))
MP_eqtl_sig_random <- read.csv(paste0(data, "MP_eqtl_sig_random.csv"))
```


# number of significant eqtl per gene
```{r}
ml_table <- ML_eqtl_sig %>% group_by(phenotype_id) %>%
  dplyr::summarise(
    count = length(variant_id))
fl_table <- FL_eqtl_sig %>% group_by(phenotype_id) %>%
  dplyr::summarise(
    count = length(variant_id))
mp_table <- MP_eqtl_sig %>% group_by(phenotype_id) %>%
  dplyr::summarise(
    count = length(variant_id))

median(ml_table$count)
median(fl_table$count)
median(mp_table$count)
```


```{r}
ggplot(ml_table, aes(count)) + geom_histogram(binwidth = 1) + theme_classic()  +theme(text = element_text(size = 12))+labs(x="Category", y="Number of eQTLs per gene")
```



# distance to TSS and number of cis-eQTLs
keep one eqtl per gene
```{r}
ml_table <- ML_eqtl_sig %>% select(start_distance)
fl_table <- FL_eqtl_sig %>% select(start_distance)
mp_table <- MP_eqtl_sig %>% select(start_distance)

ml_table <- ML_eqtl_sig_min %>% select(start_distance) 
fl_table <- FL_eqtl_sig_min %>% select(start_distance)
mp_table <- MP_eqtl_sig_min %>% select(start_distance)

ml_table <- ML_eqtl_sig_random %>% select(start_distance)
fl_table <- FL_eqtl_sig_random %>% select(start_distance)
mp_table <- MP_eqtl_sig_random %>% select(start_distance)

ml_table$category <- "ML"
fl_table$category <- "FL"
mp_table$category <- "MP"

table1 <- rbind(ml_table,fl_table,mp_table)
table2 <- rbind(ml_table,fl_table,mp_table)
table3 <- rbind(ml_table,fl_table,mp_table)


tablec <- rbind(ml_table,fl_table,mp_table)
nrow(tablec[tablec$start_distance > 0,]) # 3679 random, 3689 min, 29422 all
nrow(tablec[tablec$start_distance < 0,]) # 3659 random, 3648 min, 30944 all

test <- table2[table2$start_distance > 0,]
median(test$start_distance) # 7067 random, 7641 all, 6700 min
test <- table2[table2$start_distance < 0,]
median(test$start_distance) # -6872 random, -7782 all, -6762 min

median(abs(table1$start_distance)) # 7718
median(abs(ml_table$start_distance)) # 7601.5
median(abs(fl_table$start_distance)) # 7892.5
median(abs(mp_table$start_distance)) # 7188


p2 <- ggplot(table1, aes(x=category, y=start_distance)) + geom_violin()+ geom_boxplot(width=0.1) + theme_classic() + theme(text = element_text(size = 11))+labs(x="Tissue", y="Distance to TSS", title = "b")
p3 <- ggplot(table2, aes(x=category, y=start_distance)) + geom_violin()+ geom_boxplot(width=0.1) + theme_classic() + theme(text = element_text(size = 11))+labs(x="Tissue", y="Distance to TSS", title = "c")
p4 <- ggplot(table3, aes(x=category, y=start_distance)) + geom_violin()+ geom_boxplot(width=0.1) + theme_classic() + theme(text = element_text(size = 11))+labs(x="Tissue", y="Distance to TSS", title = "d") 
p1 + p2 + p3 + p4
```

## all cis SNPs tested
note: run code in eqtl_permutation.Rmd to get ML_eqtl etc.
```{r}
ml_table <- ML_eqtl %>% select(start_distance)
fl_table <- FL_eqtl %>% select(start_distance)
mp_table <- MP_eqtl %>% select(start_distance)
ml_table$category <- "ML"
fl_table$category <- "FL"
mp_table$category <- "MP"
table4 <- rbind(ml_table,fl_table,mp_table)

test <- table4[table4$start_distance > 0,]
median(test$start_distance) # 9061
test <- table4[table4$start_distance < 0,]
median(test$start_distance) # -9137

median(abs(table4$start_distance)) # 9099
median(abs(ml_table$start_distance)) # 8997
median(abs(fl_table$start_distance)) # 9254
median(abs(mp_table$start_distance)) # 8982


p1 <- ggplot(table4, aes(x=category, y=start_distance)) + geom_violin()+ geom_boxplot(width=0.1) + theme_classic() + theme(text = element_text(size = 11))+labs(x="Tissue", y="Distance to TSS", title = "a")

```


## look at some eGenes
note: run code in eqtl_permutation.Rmd to get ML_eqtl_full etc.
```{r}
gene1 <- ML_eqtl_full %>% filter(phenotype_id == "TX_paternal_00029650")
gene2 <- FL_eqtl_full %>% filter(phenotype_id == "TX_paternal_00030127")
gene3 <- MP_eqtl_full %>% filter(phenotype_id == "TX_paternal_00006330")

p1 <- ggplot(gene1, aes(x=start_distance, y=-log(pval_nominal))) + geom_point(alpha = 0.6) + theme_classic() + geom_hline(yintercept=-log(0.0011044), color = "red") + labs(title = "a")+ theme(text = element_text(size = 10))
p2 <- ggplot(gene2, aes(x=start_distance, y=-log(pval_nominal))) + geom_point(alpha = 0.6) + theme_classic() + geom_hline(yintercept=-log(0.00101239), color = "red") + labs(title = "b")+ theme(text = element_text(size = 10))
p3 <- ggplot(gene3, aes(x=start_distance, y=-log(pval_nominal))) + geom_point(alpha = 0.6) + theme_classic() + geom_hline(yintercept=-log(0.000283815), color = "red") + labs(title = "c")+ theme(text = element_text(size = 10))

p1 + p2 + p3 
```




# eqtls that affect multiple genes in the same tissue
## ML
```{r}
length(unique(ML_eqtl_sig$variant_id)) # 16124
ML_eqtl_multigene <- ML_eqtl_sig %>% 
     group_by(variant_id) %>%
     filter(n()>1) %>%  
    select(variant_id, maf) %>% unique() # 486

ML_eqtl_onegene <- ML_eqtl_sig %>% 
     group_by(variant_id) %>%
     filter(n()==1) %>%
    select(variant_id, maf) # 15638

median(ML_eqtl_multigene$maf) # 0.09482759
median(ML_eqtl_onegene$maf) # 0.09836066
wilcox.test(ML_eqtl_multigene$maf, ML_eqtl_onegene$maf)

p1 <- ggplot(ML_eqtl_multigene, aes(x=maf)) + geom_histogram(binwidth=0.01) + theme_classic() + ggtitle("")
p2 <- ggplot(ML_eqtl_onegene, aes(x=maf)) + geom_histogram(binwidth=0.01) + theme_classic() + ggtitle("")
p1 + p2 
```
Wilcoxon rank sum test with continuity correction

data:  ML_eqtl_multigene$maf and ML_eqtl_onegene$maf
W = 3601946, p-value = 0.04998
alternative hypothesis: true location shift is not equal to 0

## FL
```{r}
length(unique(FL_eqtl_sig$variant_id)) # 33236
FL_eqtl_multigene <- FL_eqtl_sig %>% 
     group_by(variant_id) %>%
     filter(n()>1) %>%
    select(variant_id, maf) %>% unique() # 905

FL_eqtl_onegene <- FL_eqtl_sig %>% 
     group_by(variant_id) %>%
     filter(n()==1) %>%
    select(variant_id, maf) # 32331

median(FL_eqtl_multigene$maf) # 0.1230769
median(FL_eqtl_onegene$maf) # 0.109375
wilcox.test(FL_eqtl_multigene$maf, FL_eqtl_onegene$maf)

p1 <- ggplot(FL_eqtl_multigene, aes(x=maf)) + geom_histogram(binwidth=0.01) + theme_classic() + ggtitle("")
p2 <- ggplot(FL_eqtl_onegene, aes(x=maf)) + geom_histogram(binwidth=0.01) + theme_classic() + ggtitle("")
p1 + p2 
```
Wilcoxon rank sum test with continuity correction

data:  FL_eqtl_multigene$maf and FL_eqtl_onegene$maf
W = 15610093, p-value = 0.000574
alternative hypothesis: true location shift is not equal to 0


## MP
```{r}
length(unique(MP_eqtl_sig$variant_id)) # 9251
MP_eqtl_multigene <- MP_eqtl_sig %>% 
     group_by(variant_id) %>%
     filter(n()>1) %>%
    select(variant_id, maf) %>% unique() # 201

MP_eqtl_onegene <- MP_eqtl_sig %>% 
     group_by(variant_id) %>%
     filter(n()==1) %>%
    select(variant_id, maf) # 9050

median(MP_eqtl_multigene$maf) # 0.08928572
median(MP_eqtl_onegene$maf) # 0.1034483
wilcox.test(MP_eqtl_multigene$maf, MP_eqtl_onegene$maf)

p1 <- ggplot(MP_eqtl_multigene, aes(x=maf)) + geom_histogram(binwidth=0.01) + theme_classic() + ggtitle("")
p2 <- ggplot(MP_eqtl_onegene, aes(x=maf)) + geom_histogram(binwidth=0.01) + theme_classic() + ggtitle("")
p1 + p2 
```
Wilcoxon rank sum test with continuity correction

data:  MP_eqtl_multigene\$maf and MP_eqtl_onegene$maf
W = 884480, p-value = 0.5037
alternative hypothesis: true location shift is not equal to 0

## write output
```{r}
write.csv(ML_eqtl_multigene, file = paste0(data, "ML_eqtl_multigene.csv"), row.names = F, quote = F)
write.csv(FL_eqtl_multigene, file = paste0(data, "FL_eqtl_multigene.csv"), row.names = F, quote = F)
write.csv(MP_eqtl_multigene, file = paste0(data, "MP_eqtl_multigene.csv"), row.names = F, quote = F)
```





# start distance and effect size
```{r}
ggplot(mp1, aes(x=abs(start_distance), y = abs(slope))) + geom_point()

#shapiro.test(abs(ML_eqtl_sig_min$start_distance))
cor.test(abs(ML_eqtl_sig_min$start_distance), abs(ML_eqtl_sig_min$slope), method = "spearman")
cor.test(abs(FL_eqtl_sig_min$start_distance), abs(FL_eqtl_sig_min$slope), method = "spearman")
cor.test(abs(MP_eqtl_sig_min$start_distance), abs(MP_eqtl_sig_min$slope), method = "spearman")

cor.test(abs(ML_eqtl_sig_random$start_distance), abs(ML_eqtl_sig_random$slope), method = "spearman")
cor.test(abs(FL_eqtl_sig_random$start_distance), abs(FL_eqtl_sig_random$slope), method = "spearman")
cor.test(abs(MP_eqtl_sig_random$start_distance), abs(MP_eqtl_sig_random$slope), method = "spearman")
```
Warning: Cannot compute exact p-value with ties
	Spearman's rank correlation rho

data:  abs(ML_eqtl_sig_min\$start_distance) and abs(ML_eqtl_sig_min$slope)
S = 2.463e+09, p-value = 0.04145
alternative hypothesis: true rho is not equal to 0
sample estimates:
        rho 
-0.04144365 

Warning: Cannot compute exact p-value with ties
	Spearman's rank correlation rho

data:  abs(FL_eqtl_sig_min\$start_distance) and abs(FL_eqtl_sig_min$slope)
S = 7132385773, p-value = 0.003723
alternative hypothesis: true rho is not equal to 0
sample estimates:
        rho 
-0.04943009 

Warning: Cannot compute exact p-value with ties
	Spearman's rank correlation rho

data:  abs(MP_eqtl_sig_min\$start_distance) and abs(MP_eqtl_sig_min$slope)
S = 583736272, p-value = 0.000439
alternative hypothesis: true rho is not equal to 0
sample estimates:
        rho 
-0.09142041 

Warning: Cannot compute exact p-value with ties
	Spearman's rank correlation rho

data:  abs(ML_eqtl_sig_random\$start_distance) and abs(ML_eqtl_sig_random$slope)
S = 2523692630, p-value = 0.0009553
alternative hypothesis: true rho is not equal to 0
sample estimates:
        rho 
-0.06709581 

Warning: Cannot compute exact p-value with ties
	Spearman's rank correlation rho

data:  abs(FL_eqtl_sig_random\$start_distance) and abs(FL_eqtl_sig_random$slope)
S = 7178141620, p-value = 0.0009794
alternative hypothesis: true rho is not equal to 0
sample estimates:
        rho 
-0.05616242 

Warning: Cannot compute exact p-value with ties
	Spearman's rank correlation rho

data:  abs(MP_eqtl_sig_random\$start_distance) and abs(MP_eqtl_sig_random$slope)
S = 595140826, p-value = 1.424e-05
alternative hypothesis: true rho is not equal to 0
sample estimates:
       rho 
-0.1127437

# allele frequency and effect size
```{r}
ggplot(MP_eqtl_sig_min, aes(x=maf, y = abs(slope))) + geom_point()

cor.test(ML_eqtl_sig_min$maf, abs(ML_eqtl_sig_min$slope), method = "spearman")
cor.test(FL_eqtl_sig_min$maf, abs(FL_eqtl_sig_min$slope), method = "spearman")
cor.test(MP_eqtl_sig_min$maf, abs(MP_eqtl_sig_min$slope), method = "spearman")

cor.test(ML_eqtl_sig_random$maf, abs(ML_eqtl_sig_random$slope), method = "spearman")
cor.test(FL_eqtl_sig_random$maf, abs(FL_eqtl_sig_random$slope), method = "spearman")
cor.test(MP_eqtl_sig_random$maf, abs(MP_eqtl_sig_random$slope), method = "spearman")
```
	Spearman's rank correlation rho

data:  ML_eqtl_sig_min\$maf and abs(ML_eqtl_sig_min$slope)
S = 3815138368, p-value < 2.2e-16
alternative hypothesis: true rho is not equal to 0
sample estimates:
       rho 
-0.6131593 

Warning: Cannot compute exact p-value with ties
	Spearman's rank correlation rho

data:  FL_eqtl_sig_min\$maf and abs(FL_eqtl_sig_min$slope)
S = 1.0895e+10, p-value < 2.2e-16
alternative hypothesis: true rho is not equal to 0
sample estimates:
       rho 
-0.6030069 

Spearman's rank correlation rho

data:  MP_eqtl_sig_min\$maf and abs(MP_eqtl_sig_min$slope)
S = 887735499, p-value < 2.2e-16
alternative hypothesis: true rho is not equal to 0
sample estimates:
       rho 
-0.6598123 

Spearman's rank correlation rho

data:  ML_eqtl_sig_random\$maf and abs(ML_eqtl_sig_random$slope)
S = 4055631864, p-value < 2.2e-16
alternative hypothesis: true rho is not equal to 0
sample estimates:
       rho 
-0.7148474 

Warning: Cannot compute exact p-value with ties
	Spearman's rank correlation rho

data:  FL_eqtl_sig_random\$maf and abs(FL_eqtl_sig_random$slope)
S = 1.1689e+10, p-value < 2.2e-16
alternative hypothesis: true rho is not equal to 0
sample estimates:
       rho 
-0.7198221 

Spearman's rank correlation rho

data:  MP_eqtl_sig_random\$maf and abs(MP_eqtl_sig_random$slope)
S = 934559235, p-value < 2.2e-16
alternative hypothesis: true rho is not equal to 0
sample estimates:
       rho 
-0.7473593 
