---
title: "sex interaction"
author: "Meng"
date: "2024-07-03"
output: html_document
---

```{r setup, include=FALSE}
library(arrow)
library(dplyr)
library(ggplot2)
library(patchwork)
library(tidyr)
data <- "/Users/yuanmeng/Library/CloudStorage/OneDrive-UniversityofToronto/Manuscripts/rumex_eqtl/eqtl/"

calculate_qvalues <- function(res_df, fdr = 0.1, qvalue_lambda = NULL) {
  # Number of phenotypes tested
  num_phenotypes <- nrow(res_df)
  cat(sprintf("Number of phenotypes tested: %d\n", num_phenotypes))
  
  # Determine which p-value column to use
  if ("pval_beta" %in% colnames(res_df) && !all(is.na(res_df$pval_beta))) {
    pval_col <- "pval_beta"
    correlation <- cor(res_df$pval_perm, res_df$pval_beta, use = "complete.obs")
    cat(sprintf("Correlation between Beta-approximated and empirical p-values: %.4f\n", correlation))
  } else {
    pval_col <- "pval_perm"
    cat("WARNING: no beta-approximated p-values found, using permutation p-values instead.\n")
  }
  
  # Calculate q-values
  if (!is.null(qvalue_lambda)) {
    cat(sprintf("Calculating q-values with lambda = %.3f\n", qvalue_lambda))
    qobj <- qvalue(res_df[[pval_col]], lambda = qvalue_lambda)
  } else {
    qobj <- qvalue(res_df[[pval_col]])
  }
  
  res_df$qval <- qobj$qvalues
  pi0 <- qobj$pi0
  cat(sprintf("Proportion of significant phenotypes (1-pi0): %.2f\n", 1 - pi0))
  cat(sprintf("QTL phenotypes @ FDR %.2f: %d\n", fdr, sum(res_df$qval <= fdr)))
  
  # Determine global min(p) significance threshold and calculate nominal p-value threshold for each gene
  if (pval_col == "pval_beta") {
    lb <- sort(res_df[res_df$qval <= fdr, "pval_beta"])
    ub <- sort(res_df[res_df$qval > fdr, "pval_beta"])
    
    if (length(lb) > 0) {
      lb <- tail(lb, 1)
      if (length(ub) > 0) {
        ub <- head(ub, 1)
        pthreshold <- (lb + ub) / 2
      } else {
        pthreshold <- lb
      }
      cat(sprintf("min p-value threshold @ FDR %.2f: %.6g\n", fdr, pthreshold))
      res_df$pval_nominal_threshold <- qbeta(pthreshold, res_df$beta_shape1, res_df$beta_shape2)
    }
  }
  
  return(res_df)
}

```


## permutation
/Users/yuanmeng/Library/CloudStorage/OneDrive-UniversityofToronto/Manuscripts/rumex_eqtl/eqtl/sex.cis_qtl.txt
```{r}
sex_genes <- read.table(paste0(data, "sex.cis_qtl.txt"), header = T) # 14549

hist(sex_genes$pval_beta, nclass = 20)

qval_sex <- qvalue(sex_genes$pval_beta)

summary(qval_sex)
hist(qval_sex)

```
Call:
qvalue(p = sex$pval_beta)

pi0:	0.2203196	

Cumulative number of significant calls:

          <1e-04 <0.001 <0.01 <0.025 <0.05  <0.1    <1
p-value     1707   2522  3958   4892  5895  7161 14549
q-value     1547   2391  4170   5711  7522 10216 14549
local FDR   1117   1652  2712   3505  4401  5870 14536


```{r}
sex_genes <- calculate_qvalues(sex_genes)

sex_egene <- sex_genes %>% filter(qval < 0.05) # 7522
sex_egene <- sex_genes %>% filter(qval < 0.1) # 10216
```
Number of phenotypes tested: 14549
Correlation between Beta-approximated and empirical p-values: 1.0000
Proportion of significant phenotypes (1-pi0): 0.78
QTL phenotypes @ FDR 0.05: 7522
min p-value threshold @ FDR 0.05: 0.117321


## get siginicant eQTLs without interaction

```{r}
sex_A1 <- read_parquet(paste0(data,"sex.cis_qtl_pairs.A1.parquet"))
sex_A2 <- read_parquet(paste0(data,"sex.cis_qtl_pairs.A2.parquet"))
sex_A3 <- read_parquet(paste0(data,"sex.cis_qtl_pairs.A3.parquet"))
sex_A4 <- read_parquet(paste0(data,"sex.cis_qtl_pairs.A4.parquet"))
colnames(sex_A4)
```

```{r}
sex_eqtl <- rbind(sex_A1, sex_A2, sex_A3, sex_A4) # 1331427
sum(sex_genes$num_var) # 1331427

sex_eqtl_full <- inner_join(sex_eqtl, sex_genes[,c(1,16:19)], by = "phenotype_id") # 1331427

sex_eqtl_sig <- sex_eqtl_full %>% filter(pval_nominal < pval_nominal_threshold) #  128906

sex_eqtl_sig005 <- sex_eqtl_full %>% filter(pval_nominal < pval_nominal_threshold) # 90369

sum(sex_egene$num_var) # 1026075
nrow(sex_egene) # 10216
length(unique(sex_eqtl_sig$phenotype_id)) # 10777
length(unique(sex_eqtl_sig$variant_id)) # 120718

length(unique(sex_eqtl_sig005$phenotype_id)) # 8141
length(unique(sex_eqtl_sig005$variant_id)) # 85350

```
MAF
```{r}
# MAF
sex_eqtl_sig <- sex_eqtl_sig %>% mutate(maf = ifelse(af <= 0.5, af, 1 - af))
write.csv(sex_eqtl_sig, file = paste0(data, "sex_eqtl_sig.csv"), row.names = F, quote = F)

# unique
sex_eqtl_sig_af <- ML_eqtl_sig %>% select(variant_id, af, maf) %>% distinct()

# most sigificant
ML_eqtl_sig_min <- ML_eqtl_sig %>%
  group_by(phenotype_id) %>% 
  filter(pval_nominal == min(pval_nominal)) # 3131

# random

```


## add interaction term
could be huge files, maybe just read one
```{r}
# sex_interaction_A1 <- read_parquet(paste0(data,"sex_interaction.cis_qtl_pairs.A1.parquet"))
# sex_interaction_A2 <- read_parquet(paste0(data,"sex_interaction.cis_qtl_pairs.A2.parquet"))
# sex_interaction_A3 <- read_parquet(paste0(data,"sex_interaction.cis_qtl_pairs.A3.parquet"))
sex_interaction_A4 <- read_parquet(paste0(data,"sex_interaction.cis_qtl_pairs.A4.parquet"))
colnames(sex_interaction_A4)
```

## top association
tests_emt is the effective number of independent variants in the cis-window estimated with eigenMT
```{r}
sex_inter_top <- read.table(paste0(data, "sex_interaction.cis_qtl_top_assoc.txt"), header = T) # 14374
colnames(sex_inter_top)

p_gi_001 <- sex_inter_top %>% filter(pval_gi < 0.01) # 2601

padjust_bh <- function(p, fdr_threshold = 0.1) {
  if (any(!is.finite(p))) {
    stop('P values must be finite.')
  }
  n <- length(p)
  i <- n:1
  o <- order(p, decreasing = TRUE)
  ro <- order(o)
  adjusted_p <- pmin(1, cummin(n / i * p[o]))[ro]
  
  return(adjusted_p)
}

sex_inter_top$pval_gi_adj <- padjust_bh(sex_inter_top$pval_gi, fdr_threshold = 0.1)

p_gi_adj_005 <- sex_inter_top %>% filter(pval_gi_adj < 0.05) # 235
p_gi_adj_01 <- sex_inter_top %>% filter(pval_gi_adj < 0.1) # 9253
pval_adj_bh <- sex_inter_top %>% filter(pval_adj_bh < 0.1) # 5

```

pval_g 
pval_i
pval_gi
```{r}
sex_inter_full <- inner_join(sex_inter_top, sex[,c(1,16:19)], by = "phenotype_id") # 14374
colnames(sex_inter_full)

sex_g_sig <- sex_inter_full %>% filter(pval_g < pval_nominal_threshold) # 1573
sex_i_sig <- sex_inter_full %>% filter(pval_i < pval_nominal_threshold) # 0 
sex_gi_sig <- sex_inter_full %>% filter(pval_gi < pval_nominal_threshold) # 1055
write.csv(sex_g_sig, file = paste0(data, "sex_g_sig.csv"), row.names = F, quote = F)
write.csv(sex_gi_sig, file = paste0(data, "sex_gi_sig.csv"), row.names = F, quote = F)

# check overlap
length(intersect(sex_g_sig$variant_id, sex_gi_sig$variant_id)) # 667

## MAF of eqtls
sex_g_sig <- sex_g_sig %>% mutate(maf = ifelse(af <= 0.5, af, 1 - af))
sex_gi_sig <- sex_gi_sig %>% mutate(maf = ifelse(af <= 0.5, af, 1 - af))
p1 <- ggplot(sex_g_sig, aes(x=maf)) + geom_histogram(binwidth=0.01) + theme_classic() + ggtitle("g")
p2 <- ggplot(sex_gi_sig, aes(x=maf)) + geom_histogram(binwidth=0.01) + theme_classic() + ggtitle("g x sex")
p1 + p2
```

MAF of interaction vs no interaction

p_gi_adj_01 # 9253
sex_eqtl_sig # 128906
```{r}
length(unique(sex_eqtl_sig$phenotype_id)) # 10777
length(unique(sex_eqtl_sig$variant_id)) # 120718

sex_eqtl_sig_min <- sex_eqtl_sig %>%
  group_by(phenotype_id) %>% 
  filter(pval_nominal == min(pval_nominal)) # 10784

write.table(sex_eqtl_sig_min, file = paste0(data, "sex_eqtl_sig_min.txt"), row.names = F, quote = F, sep = "\t")
sex_eqtl_sig_min


length(unique(sex_eqtl_sig_min$phenotype_id)) # 10777
length(unique(sex_eqtl_sig_min$variant_id)) # 10405

length(unique(p_gi_adj_01$phenotype_id)) # 9253
length(unique(p_gi_adj_01$variant_id)) # 9046

write.table(sex_eqtl_sig_min$variant_id, file = paste0(data, "sex_eqtl_sig_min.txt"), row.names = F, quote = F, col.names = F)
write.table(p_gi_adj_01$variant_id, file = paste0(data, "p_gi_adj_01.txt"), row.names = F, quote = F, col.names = F)
write.table(sex_eqtl_sig$variant_id, file = paste0(data, "sex_eqtl_sig.txt"), row.names = F, quote = F, col.names = F)


interaction1 <- p_gi_adj_01 %>% select(variant_id, af) %>% distinct() # 9046
all_top <- sex_eqtl_sig_min %>% select(variant_id, af)  %>% distinct() # 10784

interaction <- inner_join(interaction1, all_top, by =  "variant_id") # 368

no_interaction <-  all_top %>%
  anti_join(interaction1, by = "variant_id") # 10416

interaction <- interaction %>% mutate(maf = ifelse(af <= 0.5, af, 1 - af))
no_interaction <- no_interaction %>% mutate(maf = ifelse(af <= 0.5, af, 1 - af))


p1 <- ggplot(interaction, aes(x=maf)) + geom_histogram(binwidth=0.01) + theme_classic() + ggtitle("genotype by sex interaction")
p2 <- ggplot(no_interaction, aes(x=maf)) + geom_histogram(binwidth=0.01) + theme_classic() + ggtitle("no interaction")
p1 + p2
```

```{r}
median(interaction$maf)
IQR(interaction$maf)
median(no_interaction$maf)
IQR(no_interaction$maf)

wilcox.test(interaction$maf, no_interaction$maf)
```



## examine the eqtl with the lowest pval_gi
A1:159008371 TX_paternal_00002596
ref G alt A
```{r}
# data <- "/Users/yuanmeng/Library/CloudStorage/OneDrive-UniversityofToronto/Manuscripts/rumex_eqtl/"
normalized_counts_ln <- read.table(paste0(data,"normalized_counts_ln.bed"), header = T)
expr <- normalized_counts_ln %>% filter(gene_id == "TX_paternal_00002596")
geno <- read.table(paste0(data,"A1_159008371.vcf"), header = T)

pheno <- expr[,c(5:153)]
geno1 <- geno[,c(5:153)]
eqtl <- rbind(geno1,pheno)
eqtl <- as.data.frame(t(eqtl))

eqtl <- eqtl %>% mutate(id = row.names(eqtl))
colnames(eqtl) <- c("genotype", "phenotype", "id")

eqtl$genotype <- factor(eqtl$genotype)
levels(eqtl$genotype)
eqtl$phenotype <- as.numeric(eqtl$phenotype)

sex <- rep(NA, length(eqtl$id))
sex[grep("M", eqtl$id)] <- "M"
sex[grep("F", eqtl$id)] <- "F"
eqtl <- data.frame(eqtl, sex)

ggplot(eqtl, aes(x=genotype, y=phenotype, colour=sex)) + 
  geom_boxplot(position=position_dodge2(0.85)) + 
  geom_jitter(shape=16, position=position_jitterdodge(0.2), alpha=0.3) + 
  theme_classic()

write.table(eqtl, file = paste0(data, "A1_159008371.eqtl"), quote = F)
```


```{r}
eqtl <- eqtl %>% filter(genotype == "0/0" | genotype == "0/1")
levels(eqtl$genotype) <-c ("./.","GG", "GA","GG")

p1 <- ggplot(eqtl, aes(x=sex, y=phenotype, colour=genotype)) + 
  geom_boxplot(position=position_dodge2(0.85)) + 
  geom_jitter(shape=16, position=position_jitterdodge(0.2), alpha=0.3) + 
  theme_classic()+ theme(text = element_text(size = 15))

p2 <- ggplot(eqtl, aes(x=genotype, y=phenotype, colour=sex)) + 
  geom_boxplot(position=position_dodge2(0.85)) + 
  geom_jitter(shape=16, position=position_jitterdodge(0.2), alpha=0.3) + 
  theme_classic() + theme(text = element_text(size = 18))+ scale_colour_manual(values = c("red", "blue"))+labs(x="Genotype", y="Expression")
p2 
```



